---
title: "Differential meiotic recombination between sexes in maize"
author: "Gwonjin Lee"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: sandstone
    toc: TRUE
editor_options: 
  chunk_output_type: console
---
<div style="margin-bottom:150px;">
</div>



# 1. Recombination rate analysis for each BC1 {.tabset}

<div style="margin-bottom:50px;">
</div>

```{r, include=FALSE, message=FALSE, warning=FALSE}
# Load the required libraries
library(dplyr)
library(tidyr)
library(gplots)
library(ggplot2)
library(ggsignif)
library(stringr)
library(agricolae)
library(knitr)
#library(knitrBootstrap)
library(rmarkdown)
library(ggpubr)
library(Rmisc)
library(ggvenn)
library(reshape2)
library(VennDiagram)
library(RColorBrewer)

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage")


csv_files <- dir(pattern='*numberCO.csv$', recursive = T)


for(i in 1:length(csv_files)) {
  if(i == 1)
    COall <- read.csv(csv_files[i])
  else
    COall <- rbind(COall, read.csv(csv_files[i]))
}

colnames(COall)[1] <- c("set")

COall$sum <- rowSums(COall[,2:11]) 

COall_sum <- COall[,c(1,12)]
colnames(COall_sum) <- c("set", "CO")

COall_sum$line <- sub("F_", "_", COall_sum$set)
COall_sum$line <- sub("M_", "_", COall_sum$line)
COall_sum$line <- sub("_.*", "", COall_sum$line)
COall_sum$line <- sub("_.*", "", COall_sum$line)
COall_sum$sex <- str_sub(COall_sum$set,-5,-5)

COall_sum <- COall_sum[c("line", "sex", "set", "CO")] 
COall_sum$line <- factor(COall_sum$line, levels = c("B97", "Oh43", "Il14H", "W22", "Mo18W","Oh7B","A344", "M162W", 
                                                    "CML322", "CML69", "Ki11", "CML228", "Mo17",  "CML277", "CML103"))


# Calculate CV and SD within each sex per line
Var_CO_within <- COall_sum %>%
  dplyr::group_by(line, sex) %>%
  dplyr::summarise(
    CV = sd(CO) / mean(CO),  # Coefficient of Variation (CV)
    SD = sd(CO)  # Standard Deviation (SD)
  ) %>%
  ungroup()

Var_CO_within

mean((subset(Var_CO_within, sex == "F"))$CV)
mean((subset(Var_CO_within, sex == "M"))$CV)

# Calculate mean CO per line per sex
mean_co <- COall_sum %>%
  dplyr::group_by(line, sex) %>%
  dplyr::summarise(
    mean_CO = mean(CO)
  ) %>%
  ungroup()

# Calculate CV and SD of mean CO between lines per sex
Var_CO_sex <- mean_co %>%
  dplyr::group_by(sex) %>%
  dplyr::summarise(
    mean_CO_mean = mean(mean_CO),
    CV_mean_CO = sd(mean_CO) / mean(mean_CO),  # Coefficient of Variation (CV) of mean CO
    SD_mean_CO = sd(mean_CO)  # Standard Deviation (SD) of mean CO
  ) %>%
  ungroup()

Var_CO_sex


# Perform two-way ANOVA
anova_CO <- aov(CO ~ line + sex + line:sex, data = COall_sum) 
#interaction assesses whether the effect of sex on CO is different across different lines

# Print ANOVA summary
summary(anova_CO)

TukeyHSD(anova_CO)

CO_HSD <- HSD.test(anova_CO, c("line", "sex"), group=TRUE)
CO_HSD$groups


# Generate QQ plot
qqnorm(anova_CO$residuals)
qqline(anova_CO$residuals)

# Generate residuals vs. fitted values plot
plot(x = anova_CO$fitted.values, y = anova_CO$residuals, 
     xlab = "Fitted Values", ylab = "Residuals", main = "Residuals vs. Fitted")
abline(h = 0, col = "red")



#Mean_CO for tropical and non-tropical

mean_co <- mean_co %>%
  dplyr::mutate(
    Type = case_when(
      line %in% c("B97", "Oh43", "Il14H", "W22", "Oh7B", "A344", "Mo17", "Mo18W", "M162W") ~ "Non-tropical line",
      line %in% c("CML322", "CML69", "Ki11", "CML228", "CML277", "CML103") ~ "Tropical line",
      TRUE ~ NA_character_  # For lines not in either list
    )
  )



mean_co$TypeSex <- with(mean_co, paste(Type, sex, sep = "_"))


mean_co.aov <- aov( mean_CO ~ Type*sex, data = mean_co)
shapiro.test(mean_co.aov$residuals) #Normality?
TukeyHSD(mean_co.aov)  

mean_co_res <- HSD.test(mean_co.aov, c("Type", "sex"), group=TRUE)
#mean_co_res$groups

oder.mean_co_res <- mean_co_res$groups[order(rownames(mean_co_res$groups)),] 



Mean_CO_Trop <- summarySE(mean_co, measurevar="mean_CO", groupvars=c("Type","sex"))


################
#Bar plot


Mean_CO_Trop_p <- ggplot(Mean_CO_Trop, aes(x = Type, y = mean_CO, fill = sex)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_classic() +
   #  facet_wrap( ~ Type, nrow = 1, strip.position = "bottom", scales = "free_x" ) +
  labs(y = "Number of crossover", title = "Wilcox", x = "") +
  theme(axis.text.y = element_text(size = 12), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        strip.background = element_blank(), 
        strip.placement = "outside", 
        axis.title = element_text(size = 15), 
        strip.text = element_text(size = 14, margin = margin(0, 0, 20, 0)), 
        legend.position = "right", 
        legend.justification = "right",
        legend.title = element_blank(), 
        plot.title = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
  scale_fill_manual(values = c("red2", "blue4"),
                    breaks = c("F", "M")) +
  geom_errorbar(aes(ymin=mean_CO-sd, ymax=mean_CO+sd),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  geom_text(aes(x=Type, y=mean_CO+sd+0.8, 
                  label=c(as.character(oder.mean_co_res$groups))),
             position=position_dodge(width=0.9), size=5) 




#box

Mean_CO_Trop_p2 <- ggplot(mean_co, aes(x = Type, y = mean_CO, fill = sex)) +
  geom_boxplot(outlier.shape = NA, width = 0.9) +  # Adjust width here
  theme_classic() +
  facet_wrap(~ Type, nrow = 1, strip.position = "bottom", scales = "free_x") +
  labs(y = "Number of crossover", title = "Wilcox", x = "") +
  theme(axis.text.y = element_text(size = 12), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        strip.background = element_blank(), 
        strip.placement = "outside", 
        axis.title = element_text(size = 15), 
        strip.text = element_text(size = 14, margin = margin(0, 0, 20, 0)), 
        legend.position = "right", 
        legend.justification = "right",
        legend.title = element_blank(), 
        plot.title = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
  scale_fill_manual(values = c("red2", "blue4"),
                    breaks = c("F", "M")) +
  geom_text(data= Mean_CO_Trop, aes(x=Type, y=c(13.3,14.25,13.8,13.3), 
                  label=c(as.character(oder.mean_co_res$groups))),
             position=position_dodge(width=0.9), size=5)




#png(filename = "plot/Tro_Ntro.png", width = 5, height = 5.5, units = "in",  res = 400 )
Mean_CO_Trop_p2
#dev.off()

```



## B97
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

B97_all <- read.table("B97_mareyout.chr_snp.txt", sep = " " , header = TRUE )
B97_all$Mb <- B97_all$phys/1000000

colnames(B97_all)[2] <- "chr" 
#B97_all  <- as.data.frame(lapply(B97_all, gsub, pattern = "Chromosome ", replacement = ""))
B97_all$chr <- sub("Chromosome ", "", B97_all$chr)

B97_all$chr <- as.integer(B97_all$chr)


### combo plot
B97_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(B97_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(B97_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =B97_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="B97") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 




# Number of crossover per individual

B97_nxo_sump <- ggplot(subset(COall_sum, line == "B97"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
B97F_nxo <- read.csv("../B97/B97F_numberCO.csv", row.names = 1 ) 

B97F_nxo_df <- as.data.frame(B97F_nxo) 
B97F_nxo_df$set <- "B97F"

B97M_nxo <- read.csv("../B97/B97M_numberCO.csv", row.names = 1) 
B97M_nxo_df <- as.data.frame(B97M_nxo) 
B97M_nxo_df$set <- "B97M"

B97_nxo <- rbind(B97F_nxo_df, B97M_nxo_df)

colnames(B97_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


B97_nxo_d <- melt(B97_nxo, id = "set")
colnames(B97_nxo_d) <- c("set", "chr", "CO")


B97_nxo_sum <- summarySE(B97_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


B97_nxo_bar <- ggplot(B97_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("B97F", "B97M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

B97F_recomb_hot_genome <- read.csv("../B97/hotspot/B97F_recomb_hot_genome_0.2mb_5fold.csv" )
B97M_recomb_hot_genome <- read.csv("../B97/hotspot/B97M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in B97F_recomb_hot_genome$chr_Window
num_B97F_recomb_hot <- length(B97F_recomb_hot_genome$chr_Window)


# Get the number of rows in B97M_recomb_hot_genome$chr_Window
num_B97M_recomb_hot <- length(B97M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_B97_recomb_hot_genom <- intersect(B97M_recomb_hot_genome$chr_Window, B97F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_B97_recomb_hot <- length(common_B97_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package


B97_hotspots <- venn.diagram(
  x = list(
    B97F = B97F_recomb_hot_genome$chr_Window,
    B97M = B97M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "B97F" and "B97M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.2,   fontfamily = "arial",   
  cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(-0.4, -0.45),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_B97_recomb_hot  # Specify the size of the shared area
)




#grid.draw(B97_hotspots)


###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/B97/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
B97_phys_grouped <- B97_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
B97_phys_grouped2 <- B97_phys_grouped[!is.na(B97_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
B97_mean_recomb_perwin <- B97_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

B97_mean_recomb_perwin_pos <- B97_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
B97_mean_recomb_perwinset <- B97_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
B97_mean_recomb_perwinset$B97F[B97_mean_recomb_perwinset$B97F < 0] <- 0
B97_mean_recomb_perwinset$B97M[B97_mean_recomb_perwinset$B97M < 0] <- 0

# Genome average 
B97_genome_recomb_rate <- mean(B97_all$spline, na.rm = T )


#cutoff for general hotspots
B97_recombhotspot_cutoff <- 5 * B97_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
B97_FM_hotspot <- B97_mean_recomb_perwinset %>%
  filter((B97F > 5 * B97M | B97M > 5 * B97F) & 
         (B97F > B97_genome_recomb_rate/2 & B97M > B97_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(B97F > B97M, (B97F + 0.00001) / (B97M + 0.00001), -((B97M + 0.00001) / (B97F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(B97F > B97_recombhotspot_cutoff | B97M > B97_recombhotspot_cutoff, 
                             ifelse(B97F > B97M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(B97_FM_hotspot, "B97_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(B97_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(B97_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- B97_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- B97_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_B97 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c( "red", "blue", "pink", "skyblue"), 
                    labels = c("Female higher (hotspot)", "Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(B97_recomb_combo, 
          ggarrange(B97_nxo_bar, NA,
                    ggarrange(NA, B97_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_B97,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))



```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.

<div style="margin-bottom:150px;">
</div>


## Oh43
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

Oh43_all <- read.table("Oh43_mareyout.chr_snp.txt", sep = " " , header = TRUE )
Oh43_all$Mb <- Oh43_all$phys/1000000

colnames(Oh43_all)[2] <- "chr" 
#Oh43_all  <- as.data.frame(lapply(Oh43_all, gsub, pattern = "Chromosome ", replacement = ""))
Oh43_all$chr <- sub("Chromosome ", "", Oh43_all$chr)

Oh43_all$chr <- as.integer(Oh43_all$chr)


### combo plot
Oh43_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(Oh43_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(Oh43_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =Oh43_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="Oh43") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
Oh43_CO <- Oh43_all
Oh43_CO$chr <- as.factor(Oh43_CO$chr)

Oh43_CO_sum <- summarySE(Oh43_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


Oh43_COrate_bar <- ggplot(Oh43_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Oh43F", "Oh43M"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


Oh43_nxo_sump <- ggplot(subset(COall_sum, line == "Oh43"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



##############################################################################

## CO number per chr
Oh43F_nxo <- read.csv("../Oh43/Oh43F_numberCO.csv", row.names = 1 ) 

Oh43F_nxo_df <- as.data.frame(Oh43F_nxo) 
Oh43F_nxo_df$set <- "Oh43F"

Oh43M_nxo <- read.csv("../Oh43/Oh43M_numberCO.csv", row.names = 1) 
Oh43M_nxo_df <- as.data.frame(Oh43M_nxo) 
Oh43M_nxo_df$set <- "Oh43M"

Oh43_nxo <- rbind(Oh43F_nxo_df, Oh43M_nxo_df)

colnames(Oh43_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


Oh43_nxo_d <- melt(Oh43_nxo, id = "set")
colnames(Oh43_nxo_d) <- c("set", "chr", "CO")


Oh43_nxo_sum <- summarySE(Oh43_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


Oh43_nxo_bar <- ggplot(Oh43_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Oh43F", "Oh43M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

Oh43F_recomb_hot_genome <- read.csv("../Oh43/hotspot/Oh43F_recomb_hot_genome_0.2mb_5fold.csv" )
Oh43M_recomb_hot_genome <- read.csv("../Oh43/hotspot/Oh43M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in Oh43F_recomb_hot_genome$chr_Window
num_Oh43F_recomb_hot <- length(Oh43F_recomb_hot_genome$chr_Window)


# Get the number of rows in Oh43M_recomb_hot_genome$chr_Window
num_Oh43M_recomb_hot <- length(Oh43M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_Oh43_recomb_hot_genom <- intersect(Oh43M_recomb_hot_genome$chr_Window, Oh43F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_Oh43_recomb_hot <- length(common_Oh43_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package

Oh43_hotspots <- venn.diagram(
  x = list(
    Oh43F = Oh43F_recomb_hot_genome$chr_Window,
    Oh43M = Oh43M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "Oh43F" and "Oh43M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(-0.3, -0.45),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_Oh43_recomb_hot  # Specify the size of the shared area
)


#grid.draw(Oh43_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/Oh43/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
Oh43_phys_grouped <- Oh43_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
Oh43_phys_grouped2 <- Oh43_phys_grouped[!is.na(Oh43_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
Oh43_mean_recomb_perwin <- Oh43_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

Oh43_mean_recomb_perwin_pos <- Oh43_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
Oh43_mean_recomb_perwinset <- Oh43_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
Oh43_mean_recomb_perwinset$Oh43F[Oh43_mean_recomb_perwinset$Oh43F < 0] <- 0
Oh43_mean_recomb_perwinset$Oh43M[Oh43_mean_recomb_perwinset$Oh43M < 0] <- 0

# Genome average 
Oh43_genome_recomb_rate <- mean(Oh43_all$spline, na.rm = T )


#cutoff for general hotspots
Oh43_recombhotspot_cutoff <- 5 * Oh43_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
Oh43_FM_hotspot <- Oh43_mean_recomb_perwinset %>%
  filter((Oh43F > 5 * Oh43M | Oh43M > 5 * Oh43F) & 
         (Oh43F > Oh43_genome_recomb_rate/2 & Oh43M > Oh43_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(Oh43F > Oh43M, (Oh43F + 0.00001) / (Oh43M + 0.00001), -((Oh43M + 0.00001) / (Oh43F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(Oh43F > Oh43_recombhotspot_cutoff | Oh43M > Oh43_recombhotspot_cutoff, 
                             ifelse(Oh43F > Oh43M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(Oh43_FM_hotspot, "Oh43_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(Oh43_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(Oh43_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- Oh43_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- Oh43_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_Oh43 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
   scale_fill_manual(values = c( "blue", "pink",  "skyblue"),
                    labels = c( "Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(Oh43_recomb_combo, 
          ggarrange(Oh43_nxo_bar, NA,
                    ggarrange(NA, Oh43_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_Oh43,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))



```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>

## Il14H
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

Il14H_all <- read.table("Il14H_mareyout.chr_snp.txt", sep = " " , header = TRUE )
Il14H_all$Mb <- Il14H_all$phys/1000000

colnames(Il14H_all)[2] <- "chr" 
#Il14H_all  <- as.data.frame(lapply(Il14H_all, gsub, pattern = "Chromosome ", replacement = ""))
Il14H_all$chr <- sub("Chromosome ", "", Il14H_all$chr)

Il14H_all$chr <- as.integer(Il14H_all$chr)


### combo plot
Il14H_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(Il14H_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(Il14H_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =Il14H_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="Il14H") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
Il14H_CO <- Il14H_all
Il14H_CO$chr <- as.factor(Il14H_CO$chr)

Il14H_CO_sum <- summarySE(Il14H_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


Il14H_COrate_bar <- ggplot(Il14H_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Il14HF", "Il14HM"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


Il14H_nxo_sump <- ggplot(subset(COall_sum, line == "Il14H"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

##############################################################################

## CO number per chr
Il14HF_nxo <- read.csv("../Il14H/Il14HF_numberCO.csv", row.names = 1 ) 

Il14HF_nxo_df <- as.data.frame(Il14HF_nxo) 
Il14HF_nxo_df$set <- "Il14HF"

Il14HM_nxo <- read.csv("../Il14H/Il14HM_numberCO.csv", row.names = 1) 
Il14HM_nxo_df <- as.data.frame(Il14HM_nxo) 
Il14HM_nxo_df$set <- "Il14HM"

Il14H_nxo <- rbind(Il14HF_nxo_df, Il14HM_nxo_df)

colnames(Il14H_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


Il14H_nxo_d <- melt(Il14H_nxo, id = "set")
colnames(Il14H_nxo_d) <- c("set", "chr", "CO")


Il14H_nxo_sum <- summarySE(Il14H_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


Il14H_nxo_bar <- ggplot(Il14H_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Il14HF", "Il14HM"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

Il14HF_recomb_hot_genome <- read.csv("../Il14H/hotspot/Il14HF_recomb_hot_genome_0.2mb_5fold.csv" )
Il14HM_recomb_hot_genome <- read.csv("../Il14H/hotspot/Il14HM_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in Il14HF_recomb_hot_genome$chr_Window
num_Il14HF_recomb_hot <- length(Il14HF_recomb_hot_genome$chr_Window)


# Get the number of rows in Il14HM_recomb_hot_genome$chr_Window
num_Il14HM_recomb_hot <- length(Il14HM_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_Il14H_recomb_hot_genom <- intersect(Il14HM_recomb_hot_genome$chr_Window, Il14HF_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_Il14H_recomb_hot <- length(common_Il14H_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package
Il14H_hotspots <- venn.diagram(
  x = list(
    Il14HF = Il14HF_recomb_hot_genome$chr_Window,
    Il14HM = Il14HM_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "Il14HF" and "Il14HM"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(-0.4, -0.45),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_Il14H_recomb_hot  # Specify the size of the shared area
)


#grid.draw(Il14H_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/Il14H/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
Il14H_phys_grouped <- Il14H_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
Il14H_phys_grouped2 <- Il14H_phys_grouped[!is.na(Il14H_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
Il14H_mean_recomb_perwin <- Il14H_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

Il14H_mean_recomb_perwin_pos <- Il14H_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
Il14H_mean_recomb_perwinset <- Il14H_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
Il14H_mean_recomb_perwinset$Il14HF[Il14H_mean_recomb_perwinset$Il14HF < 0] <- 0
Il14H_mean_recomb_perwinset$Il14HM[Il14H_mean_recomb_perwinset$Il14HM < 0] <- 0

# Genome average 
Il14H_genome_recomb_rate <- mean(Il14H_all$spline, na.rm = T )


#cutoff for general hotspots
Il14H_recombhotspot_cutoff <- 5 * Il14H_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
Il14H_FM_hotspot <- Il14H_mean_recomb_perwinset %>%
  filter((Il14HF > 5 * Il14HM | Il14HM > 5 * Il14HF) & 
         (Il14HF > Il14H_genome_recomb_rate/2 & Il14HM > Il14H_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(Il14HF > Il14HM, (Il14HF + 0.00001) / (Il14HM + 0.00001), -((Il14HM + 0.00001) / (Il14HF + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(Il14HF > Il14H_recombhotspot_cutoff | Il14HM > Il14H_recombhotspot_cutoff, 
                             ifelse(Il14HF > Il14HM, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(Il14H_FM_hotspot, "Il14H_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(Il14H_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(Il14H_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- Il14H_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- Il14H_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_Il14H <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c( "blue", "skyblue"), 
                    labels = c( "Male higher (hotspot)",  "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 



# Merge plots
ggarrange(Il14H_recomb_combo, 
          ggarrange(Il14H_nxo_bar, NA,
                    ggarrange(NA, Il14H_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_Il14H,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))



```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>

## W22
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

W22_all <- read.table("W22_mareyout.chr_snp.txt", sep = " " , header = TRUE )
W22_all$Mb <- W22_all$phys/1000000

colnames(W22_all)[2] <- "chr" 
#W22_all  <- as.data.frame(lapply(W22_all, gsub, pattern = "Chromosome ", replacement = ""))
W22_all$chr <- sub("Chromosome ", "", W22_all$chr)

W22_all$chr <- as.integer(W22_all$chr)


### combo plot
W22_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(W22_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(W22_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =W22_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="W22") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
W22_CO <- W22_all
W22_CO$chr <- as.factor(W22_CO$chr)

W22_CO_sum <- summarySE(W22_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


W22_COrate_bar <- ggplot(W22_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("W22F", "W22M"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


W22_nxo_sump <- ggplot(subset(COall_sum, line == "W22"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
W22F_nxo <- read.csv("../W22/W22F_numberCO.csv", row.names = 1 ) 

W22F_nxo_df <- as.data.frame(W22F_nxo) 
W22F_nxo_df$set <- "W22F"

W22M_nxo <- read.csv("../W22/W22M_numberCO.csv", row.names = 1) 
W22M_nxo_df <- as.data.frame(W22M_nxo) 
W22M_nxo_df$set <- "W22M"

W22_nxo <- rbind(W22F_nxo_df, W22M_nxo_df)

colnames(W22_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


W22_nxo_d <- melt(W22_nxo, id = "set")
colnames(W22_nxo_d) <- c("set", "chr", "CO")


W22_nxo_sum <- summarySE(W22_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


W22_nxo_bar <- ggplot(W22_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("W22F", "W22M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

W22F_recomb_hot_genome <- read.csv("../W22/hotspot/W22F_recomb_hot_genome_0.2mb_5fold.csv" )
W22M_recomb_hot_genome <- read.csv("../W22/hotspot/W22M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in W22F_recomb_hot_genome$chr_Window
num_W22F_recomb_hot <- length(W22F_recomb_hot_genome$chr_Window)


# Get the number of rows in W22M_recomb_hot_genome$chr_Window
num_W22M_recomb_hot <- length(W22M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_W22_recomb_hot_genom <- intersect(W22M_recomb_hot_genome$chr_Window, W22F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_W22_recomb_hot <- length(common_W22_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package
W22_hotspots <- venn.diagram(
  x = list(
    W22F = W22F_recomb_hot_genome$chr_Window,
    W22M = W22M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "W22F" and "W22M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(-0.32, -0.43),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_W22_recomb_hot  # Specify the size of the shared area
)


#grid.draw(W22_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/W22/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
W22_phys_grouped <- W22_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
W22_phys_grouped2 <- W22_phys_grouped[!is.na(W22_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
W22_mean_recomb_perwin <- W22_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

W22_mean_recomb_perwin_pos <- W22_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
W22_mean_recomb_perwinset <- W22_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
W22_mean_recomb_perwinset$W22F[W22_mean_recomb_perwinset$W22F < 0] <- 0
W22_mean_recomb_perwinset$W22M[W22_mean_recomb_perwinset$W22M < 0] <- 0

# Genome average 
W22_genome_recomb_rate <- mean(W22_all$spline, na.rm = T )


#cutoff for general hotspots
W22_recombhotspot_cutoff <- 5 * W22_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
W22_FM_hotspot <- W22_mean_recomb_perwinset %>%
  filter((W22F > 5 * W22M | W22M > 5 * W22F) & 
         (W22F > W22_genome_recomb_rate/2 & W22M > W22_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(W22F > W22M, (W22F + 0.00001) / (W22M + 0.00001), -((W22M + 0.00001) / (W22F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(W22F > W22_recombhotspot_cutoff | W22M > W22_recombhotspot_cutoff, 
                             ifelse(W22F > W22M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(W22_FM_hotspot, "W22_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(W22_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(W22_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- W22_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- W22_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_W22 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
   scale_fill_manual(values = c( "blue", "pink",  "skyblue"),
                    labels = c("Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(W22_recomb_combo, 
          ggarrange(W22_nxo_bar, NA,
                    ggarrange(NA, W22_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_W22,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))



```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>


## Mo18W
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

Mo18W_all <- read.table("Mo18W_mareyout.chr_snp.txt", sep = " " , header = TRUE )
Mo18W_all$Mb <- Mo18W_all$phys/1000000

colnames(Mo18W_all)[2] <- "chr" 
#Mo18W_all  <- as.data.frame(lapply(Mo18W_all, gsub, pattern = "Chromosome ", replacement = ""))
Mo18W_all$chr <- sub("Chromosome ", "", Mo18W_all$chr)

Mo18W_all$chr <- as.integer(Mo18W_all$chr)


### combo plot
Mo18W_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(Mo18W_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(Mo18W_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =Mo18W_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="Mo18W") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
Mo18W_CO <- Mo18W_all
Mo18W_CO$chr <- as.factor(Mo18W_CO$chr)

Mo18W_CO_sum <- summarySE(Mo18W_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


Mo18W_COrate_bar <- ggplot(Mo18W_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Mo18WF", "Mo18WM"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


Mo18W_nxo_sump <- ggplot(subset(COall_sum, line == "Mo18W"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
Mo18WF_nxo <- read.csv("../Mo18W/Mo18WF_numberCO.csv", row.names = 1 ) 

Mo18WF_nxo_df <- as.data.frame(Mo18WF_nxo) 
Mo18WF_nxo_df$set <- "Mo18WF"

Mo18WM_nxo <- read.csv("../Mo18W/Mo18WM_numberCO.csv", row.names = 1) 
Mo18WM_nxo_df <- as.data.frame(Mo18WM_nxo) 
Mo18WM_nxo_df$set <- "Mo18WM"

Mo18W_nxo <- rbind(Mo18WF_nxo_df, Mo18WM_nxo_df)

colnames(Mo18W_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


Mo18W_nxo_d <- melt(Mo18W_nxo, id = "set")
colnames(Mo18W_nxo_d) <- c("set", "chr", "CO")


Mo18W_nxo_sum <- summarySE(Mo18W_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


Mo18W_nxo_bar <- ggplot(Mo18W_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Mo18WF", "Mo18WM"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

Mo18WF_recomb_hot_genome <- read.csv("../Mo18W/hotspot/Mo18WF_recomb_hot_genome_0.2mb_5fold.csv" )
Mo18WM_recomb_hot_genome <- read.csv("../Mo18W/hotspot/Mo18WM_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in Mo18WF_recomb_hot_genome$chr_Window
num_Mo18WF_recomb_hot <- length(Mo18WF_recomb_hot_genome$chr_Window)


# Get the number of rows in Mo18WM_recomb_hot_genome$chr_Window
num_Mo18WM_recomb_hot <- length(Mo18WM_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_Mo18W_recomb_hot_genom <- intersect(Mo18WM_recomb_hot_genome$chr_Window, Mo18WF_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_Mo18W_recomb_hot <- length(common_Mo18W_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package

Mo18W_hotspots <- venn.diagram(
  x = list(
    Mo18WF = Mo18WF_recomb_hot_genome$chr_Window,
    Mo18WM = Mo18WM_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "Mo18WF" and "Mo18WM"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(-0.41, -0.44),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_Mo18W_recomb_hot  # Specify the size of the shared area
)


#grid.draw(Mo18W_hotspots)


###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/Mo18W/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
Mo18W_phys_grouped <- Mo18W_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
Mo18W_phys_grouped2 <- Mo18W_phys_grouped[!is.na(Mo18W_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
Mo18W_mean_recomb_perwin <- Mo18W_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

Mo18W_mean_recomb_perwin_pos <- Mo18W_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
Mo18W_mean_recomb_perwinset <- Mo18W_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
Mo18W_mean_recomb_perwinset$Mo18WF[Mo18W_mean_recomb_perwinset$Mo18WF < 0] <- 0
Mo18W_mean_recomb_perwinset$Mo18WM[Mo18W_mean_recomb_perwinset$Mo18WM < 0] <- 0

# Genome average 
Mo18W_genome_recomb_rate <- mean(Mo18W_all$spline, na.rm = T )


#cutoff for general hotspots

Mo18W_recombhotspot_cutoff <- 5 * Mo18W_genome_recomb_rate  #5-fold higher than genome average


# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
Mo18W_FM_hotspot <- Mo18W_mean_recomb_perwinset %>%
  filter((Mo18WF > 5 * Mo18WM | Mo18WM > 5 * Mo18WF) & 
         (Mo18WF > Mo18W_genome_recomb_rate/2 & Mo18WM > Mo18W_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(Mo18WF > Mo18WM, (Mo18WF + 0.00001) / (Mo18WM + 0.00001), -((Mo18WM + 0.00001) / (Mo18WF + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(Mo18WF > Mo18W_recombhotspot_cutoff | Mo18WM > Mo18W_recombhotspot_cutoff, 
                             ifelse(Mo18WF > Mo18WM, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(Mo18W_FM_hotspot, "Mo18W_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(Mo18W_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(Mo18W_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- Mo18W_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- Mo18W_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_Mo18W <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
   scale_fill_manual(values = c( "blue", "pink",  "skyblue"),
                    labels = c("Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 



# Merge plots
ggarrange(Mo18W_recomb_combo, 
          ggarrange(Mo18W_nxo_bar, NA,
                    ggarrange(NA, Mo18W_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_Mo18W,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))



```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>


## Oh7B
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

Oh7B_all <- read.table("Oh7B_mareyout.chr_snp.txt", sep = " " , header = TRUE )
Oh7B_all$Mb <- Oh7B_all$phys/1000000

colnames(Oh7B_all)[2] <- "chr" 
#Oh7B_all  <- as.data.frame(lapply(Oh7B_all, gsub, pattern = "Chromosome ", replacement = ""))
Oh7B_all$chr <- sub("Chromosome ", "", Oh7B_all$chr)

Oh7B_all$chr <- as.integer(Oh7B_all$chr)


### combo plot
Oh7B_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(Oh7B_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(Oh7B_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =Oh7B_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="Oh7B") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
Oh7B_CO <- Oh7B_all
Oh7B_CO$chr <- as.factor(Oh7B_CO$chr)

Oh7B_CO_sum <- summarySE(Oh7B_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


Oh7B_COrate_bar <- ggplot(Oh7B_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Oh7BF", "Oh7BM"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


Oh7B_nxo_sump <- ggplot(subset(COall_sum, line == "Oh7B"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
Oh7BF_nxo <- read.csv("../Oh7B/Oh7BF_numberCO.csv", row.names = 1 ) 

Oh7BF_nxo_df <- as.data.frame(Oh7BF_nxo) 
Oh7BF_nxo_df$set <- "Oh7BF"

Oh7BM_nxo <- read.csv("../Oh7B/Oh7BM_numberCO.csv", row.names = 1) 
Oh7BM_nxo_df <- as.data.frame(Oh7BM_nxo) 
Oh7BM_nxo_df$set <- "Oh7BM"

Oh7B_nxo <- rbind(Oh7BF_nxo_df, Oh7BM_nxo_df)

colnames(Oh7B_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


Oh7B_nxo_d <- melt(Oh7B_nxo, id = "set")
colnames(Oh7B_nxo_d) <- c("set", "chr", "CO")


Oh7B_nxo_sum <- summarySE(Oh7B_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


Oh7B_nxo_bar <- ggplot(Oh7B_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Oh7BF", "Oh7BM"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

Oh7BF_recomb_hot_genome <- read.csv("../Oh7B/hotspot/Oh7BF_recomb_hot_genome_0.2mb_5fold.csv" )
Oh7BM_recomb_hot_genome <- read.csv("../Oh7B/hotspot/Oh7BM_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in Oh7BF_recomb_hot_genome$chr_Window
num_Oh7BF_recomb_hot <- length(Oh7BF_recomb_hot_genome$chr_Window)


# Get the number of rows in Oh7BM_recomb_hot_genome$chr_Window
num_Oh7BM_recomb_hot <- length(Oh7BM_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_Oh7B_recomb_hot_genom <- intersect(Oh7BM_recomb_hot_genome$chr_Window, Oh7BF_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_Oh7B_recomb_hot <- length(common_Oh7B_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package


Oh7B_hotspots <- venn.diagram(
  x = list(
    Oh7BF = Oh7BF_recomb_hot_genome$chr_Window,
    Oh7BM = Oh7BM_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "Oh7BF" and "Oh7BM"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(0.045, 0.045),# Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  shared = num_common_Oh7B_recomb_hot  # Specify the size of the shared area
)


#grid.draw(Oh7B_hotspots)

###

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/Oh7B/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
Oh7B_phys_grouped <- Oh7B_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
Oh7B_phys_grouped2 <- Oh7B_phys_grouped[!is.na(Oh7B_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
Oh7B_mean_recomb_perwin <- Oh7B_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

Oh7B_mean_recomb_perwin_pos <- Oh7B_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
Oh7B_mean_recomb_perwinset <- Oh7B_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
Oh7B_mean_recomb_perwinset$Oh7BF[Oh7B_mean_recomb_perwinset$Oh7BF < 0] <- 0
Oh7B_mean_recomb_perwinset$Oh7BM[Oh7B_mean_recomb_perwinset$Oh7BM < 0] <- 0

# Genome average 
Oh7B_genome_recomb_rate <- mean(Oh7B_all$spline, na.rm = T )


#cutoff for general hotspots
Oh7B_recombhotspot_cutoff <- 5 * Oh7B_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
Oh7B_FM_hotspot <- Oh7B_mean_recomb_perwinset %>%
  filter((Oh7BF > 5 * Oh7BM | Oh7BM > 5 * Oh7BF) & 
         (Oh7BF > Oh7B_genome_recomb_rate/2 & Oh7BM > Oh7B_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(Oh7BF > Oh7BM, (Oh7BF + 0.00001) / (Oh7BM + 0.00001), -((Oh7BM + 0.00001) / (Oh7BF + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(Oh7BF > Oh7B_recombhotspot_cutoff | Oh7BM > Oh7B_recombhotspot_cutoff, 
                             ifelse(Oh7BF > Oh7BM, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(Oh7B_FM_hotspot, "Oh7B_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(Oh7B_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(Oh7B_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- Oh7B_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- Oh7B_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_Oh7B <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c( "red",  "pink", "skyblue"), 
                    labels = c("Female higher (hotspot)",  "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(Oh7B_recomb_combo, 
          ggarrange(Oh7B_nxo_bar, NA,
                    ggarrange(NA, Oh7B_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_Oh7B,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))



```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>



## A344
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

A344_all <- read.table("A344_mareyout.chr_snp.txt", sep = " " , header = TRUE )
A344_all$Mb <- A344_all$phys/1000000

colnames(A344_all)[2] <- "chr" 
#A344_all  <- as.data.frame(lapply(A344_all, gsub, pattern = "Chromosome ", replacement = ""))
A344_all$chr <- sub("Chromosome ", "", A344_all$chr)

A344_all$chr <- as.integer(A344_all$chr)


### combo plot
A344_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(A344_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(A344_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =A344_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="A344") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
A344_CO <- A344_all
A344_CO$chr <- as.factor(A344_CO$chr)

A344_CO_sum <- summarySE(A344_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


A344_COrate_bar <- ggplot(A344_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("A344F", "A344M"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


A344_nxo_sump <- ggplot(subset(COall_sum, line == "A344"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
A344F_nxo <- read.csv("../A344/A344F_numberCO.csv", row.names = 1 ) 

A344F_nxo_df <- as.data.frame(A344F_nxo) 
A344F_nxo_df$set <- "A344F"

A344M_nxo <- read.csv("../A344/A344M_numberCO.csv", row.names = 1) 
A344M_nxo_df <- as.data.frame(A344M_nxo) 
A344M_nxo_df$set <- "A344M"

A344_nxo <- rbind(A344F_nxo_df, A344M_nxo_df)

colnames(A344_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


A344_nxo_d <- melt(A344_nxo, id = "set")
colnames(A344_nxo_d) <- c("set", "chr", "CO")


A344_nxo_sum <- summarySE(A344_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


A344_nxo_bar <- ggplot(A344_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("A344F", "A344M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

A344F_recomb_hot_genome <- read.csv("../A344/hotspot/A344F_recomb_hot_genome_0.2mb_5fold.csv" )
A344M_recomb_hot_genome <- read.csv("../A344/hotspot/A344M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in A344F_recomb_hot_genome$chr_Window
num_A344F_recomb_hot <- length(A344F_recomb_hot_genome$chr_Window)


# Get the number of rows in A344M_recomb_hot_genome$chr_Window
num_A344M_recomb_hot <- length(A344M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_A344_recomb_hot_genom <- intersect(A344M_recomb_hot_genome$chr_Window, A344F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_A344_recomb_hot <- length(common_A344_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package

A344_hotspots <- venn.diagram(
  x = list(
    A344F = A344F_recomb_hot_genome$chr_Window,
    A344M = A344M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "A344F" and "A344M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(0.05, 0.05),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_A344_recomb_hot  # Specify the size of the shared area
)


#grid.draw(A344_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/A344/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
A344_phys_grouped <- A344_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
A344_phys_grouped2 <- A344_phys_grouped[!is.na(A344_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
A344_mean_recomb_perwin <- A344_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

A344_mean_recomb_perwin_pos <- A344_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
A344_mean_recomb_perwinset <- A344_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
A344_mean_recomb_perwinset$A344F[A344_mean_recomb_perwinset$A344F < 0] <- 0
A344_mean_recomb_perwinset$A344M[A344_mean_recomb_perwinset$A344M < 0] <- 0

# Genome average 
A344_genome_recomb_rate <- mean(A344_all$spline, na.rm = T )


#cutoff for general hotspots
A344_recombhotspot_cutoff <- 5 * A344_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
A344_FM_hotspot <- A344_mean_recomb_perwinset %>%
  filter((A344F > 5 * A344M | A344M > 5 * A344F) & 
         (A344F > A344_genome_recomb_rate/2 & A344M > A344_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(A344F > A344M, (A344F + 0.00001) / (A344M + 0.00001), -((A344M + 0.00001) / (A344F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(A344F > A344_recombhotspot_cutoff | A344M > A344_recombhotspot_cutoff, 
                             ifelse(A344F > A344M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(A344_FM_hotspot, "A344_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(A344_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(A344_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- A344_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- A344_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_A344 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c(  "pink", "skyblue"), 
                    labels = c("Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(A344_recomb_combo, 
          ggarrange(A344_nxo_bar, NA,
                    ggarrange(NA, A344_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_A344,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))



```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>



## M162W
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

M162W_all <- read.table("M162W_mareyout.chr_snp.txt", sep = " " , header = TRUE )
M162W_all$Mb <- M162W_all$phys/1000000

colnames(M162W_all)[2] <- "chr" 
#M162W_all  <- as.data.frame(lapply(M162W_all, gsub, pattern = "Chromosome ", replacement = ""))
M162W_all$chr <- sub("Chromosome ", "", M162W_all$chr)

M162W_all$chr <- as.integer(M162W_all$chr)


### combo plot
M162W_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(M162W_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(M162W_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =M162W_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="M162W") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
M162W_CO <- M162W_all
M162W_CO$chr <- as.factor(M162W_CO$chr)

M162W_CO_sum <- summarySE(M162W_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


M162W_COrate_bar <- ggplot(M162W_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("M162WF", "M162WM"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


M162W_nxo_sump <- ggplot(subset(COall_sum, line == "M162W"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
M162WF_nxo <- read.csv("../M162W/M162WF_numberCO.csv", row.names = 1 ) 

M162WF_nxo_df <- as.data.frame(M162WF_nxo) 
M162WF_nxo_df$set <- "M162WF"

M162WM_nxo <- read.csv("../M162W/M162WM_numberCO.csv", row.names = 1) 
M162WM_nxo_df <- as.data.frame(M162WM_nxo) 
M162WM_nxo_df$set <- "M162WM"

M162W_nxo <- rbind(M162WF_nxo_df, M162WM_nxo_df)

colnames(M162W_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


M162W_nxo_d <- melt(M162W_nxo, id = "set")
colnames(M162W_nxo_d) <- c("set", "chr", "CO")


M162W_nxo_sum <- summarySE(M162W_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


M162W_nxo_bar <- ggplot(M162W_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("M162WF", "M162WM"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

M162WF_recomb_hot_genome <- read.csv("../M162W/hotspot/M162WF_recomb_hot_genome_0.2mb_5fold.csv" )
M162WM_recomb_hot_genome <- read.csv("../M162W/hotspot/M162WM_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in M162WF_recomb_hot_genome$chr_Window
num_M162WF_recomb_hot <- length(M162WF_recomb_hot_genome$chr_Window)


# Get the number of rows in M162WM_recomb_hot_genome$chr_Window
num_M162WM_recomb_hot <- length(M162WM_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_M162W_recomb_hot_genom <- intersect(M162WM_recomb_hot_genome$chr_Window, M162WF_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_M162W_recomb_hot <- length(common_M162W_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package

M162W_hotspots <- venn.diagram(
  x = list(
    M162WF = M162WF_recomb_hot_genome$chr_Window,
    M162WM = M162WM_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "M162WF" and "M162WM"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(0.035, 0.035),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_M162W_recomb_hot  # Specify the size of the shared area
)


#grid.draw(M162W_hotspots)
###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/M162W/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
M162W_phys_grouped <- M162W_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
M162W_phys_grouped2 <- M162W_phys_grouped[!is.na(M162W_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
M162W_mean_recomb_perwin <- M162W_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

M162W_mean_recomb_perwin_pos <- M162W_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
M162W_mean_recomb_perwinset <- M162W_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
M162W_mean_recomb_perwinset$M162WF[M162W_mean_recomb_perwinset$M162WF < 0] <- 0
M162W_mean_recomb_perwinset$M162WM[M162W_mean_recomb_perwinset$M162WM < 0] <- 0

# Genome average 
M162W_genome_recomb_rate <- mean(M162W_all$spline, na.rm = T )


#cutoff for general hotspots
M162W_recombhotspot_cutoff <- 5 * M162W_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
M162W_FM_hotspot <- M162W_mean_recomb_perwinset %>%
  filter((M162WF > 5 * M162WM | M162WM > 5 * M162WF) & 
         (M162WF > M162W_genome_recomb_rate/2 & M162WM > M162W_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(M162WF > M162WM, (M162WF + 0.00001) / (M162WM + 0.00001), -((M162WM + 0.00001) / (M162WF + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(M162WF > M162W_recombhotspot_cutoff | M162WM > M162W_recombhotspot_cutoff, 
                             ifelse(M162WF > M162WM, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(M162W_FM_hotspot, "M162W_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(M162W_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(M162W_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- M162W_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- M162W_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_M162W <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c( "red", "blue", "pink", "skyblue"), 
                    labels = c("Female higher (hotspot)", "Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(M162W_recomb_combo, 
          ggarrange(M162W_nxo_bar, NA,
                    ggarrange(NA, M162W_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_M162W,ncol =5, widths = c(1.3,0.07, 0.6, 0.07, 0.77)), 
          nrow = 2, heights = c(1,0.6))



```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>




## CML322
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

CML322_all <- read.table("CML322_mareyout.chr_snp.txt", sep = " " , header = TRUE )
CML322_all$Mb <- CML322_all$phys/1000000

colnames(CML322_all)[2] <- "chr" 
#CML322_all  <- as.data.frame(lapply(CML322_all, gsub, pattern = "Chromosome ", replacement = ""))
CML322_all$chr <- sub("Chromosome ", "", CML322_all$chr)

CML322_all$chr <- as.integer(CML322_all$chr)


### combo plot
CML322_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(CML322_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(CML322_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =CML322_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="CML322") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
CML322_CO <- CML322_all
CML322_CO$chr <- as.factor(CML322_CO$chr)

CML322_CO_sum <- summarySE(CML322_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


CML322_COrate_bar <- ggplot(CML322_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("CML322F", "CML322M"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


CML322_nxo_sump <- ggplot(subset(COall_sum, line == "CML322"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
CML322F_nxo <- read.csv("../CML322/CML322F_numberCO.csv", row.names = 1 ) 

CML322F_nxo_df <- as.data.frame(CML322F_nxo) 
CML322F_nxo_df$set <- "CML322F"

CML322M_nxo <- read.csv("../CML322/CML322M_numberCO.csv", row.names = 1) 
CML322M_nxo_df <- as.data.frame(CML322M_nxo) 
CML322M_nxo_df$set <- "CML322M"

CML322_nxo <- rbind(CML322F_nxo_df, CML322M_nxo_df)

colnames(CML322_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


CML322_nxo_d <- melt(CML322_nxo, id = "set")
colnames(CML322_nxo_d) <- c("set", "chr", "CO")


CML322_nxo_sum <- summarySE(CML322_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


CML322_nxo_bar <- ggplot(CML322_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("CML322F", "CML322M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

CML322F_recomb_hot_genome <- read.csv("../CML322/hotspot/CML322F_recomb_hot_genome_0.2mb_5fold.csv" )
CML322M_recomb_hot_genome <- read.csv("../CML322/hotspot/CML322M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in CML322F_recomb_hot_genome$chr_Window
num_CML322F_recomb_hot <- length(CML322F_recomb_hot_genome$chr_Window)


# Get the number of rows in CML322M_recomb_hot_genome$chr_Window
num_CML322M_recomb_hot <- length(CML322M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_CML322_recomb_hot_genom <- intersect(CML322M_recomb_hot_genome$chr_Window, CML322F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_CML322_recomb_hot <- length(common_CML322_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package


CML322_hotspots <- venn.diagram(
  x = list(
    CML322F = CML322F_recomb_hot_genome$chr_Window,
    CML322M = CML322M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "CML322F" and "CML322M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(-0.37, -0.45),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_CML322_recomb_hot  # Specify the size of the shared area
)


#grid.draw(CML322_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/CML322/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
CML322_phys_grouped <- CML322_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
CML322_phys_grouped2 <- CML322_phys_grouped[!is.na(CML322_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
CML322_mean_recomb_perwin <- CML322_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

CML322_mean_recomb_perwin_pos <- CML322_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
CML322_mean_recomb_perwinset <- CML322_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
CML322_mean_recomb_perwinset$CML322F[CML322_mean_recomb_perwinset$CML322F < 0] <- 0
CML322_mean_recomb_perwinset$CML322M[CML322_mean_recomb_perwinset$CML322M < 0] <- 0

# Genome average 
CML322_genome_recomb_rate <- mean(CML322_all$spline, na.rm = T )


#cutoff for general hotspots
CML322_recombhotspot_cutoff <- 5 * CML322_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
CML322_FM_hotspot <- CML322_mean_recomb_perwinset %>%
  filter((CML322F > 5 * CML322M | CML322M > 5 * CML322F) & 
         (CML322F > CML322_genome_recomb_rate/2 & CML322M > CML322_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(CML322F > CML322M, (CML322F + 0.00001) / (CML322M + 0.00001), -((CML322M + 0.00001) / (CML322F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(CML322F > CML322_recombhotspot_cutoff | CML322M > CML322_recombhotspot_cutoff, 
                             ifelse(CML322F > CML322M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(CML322_FM_hotspot, "CML322_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(CML322_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(CML322_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- CML322_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- CML322_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_CML322 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c( "blue", "pink", "skyblue"), 
                    labels = c("Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(CML322_recomb_combo, 
          ggarrange(CML322_nxo_bar, NA,
                    ggarrange(NA, CML322_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_CML322,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))




```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>

## CML69
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

CML69_all <- read.table("CML69_mareyout.chr_snp.txt", sep = " " , header = TRUE )
CML69_all$Mb <- CML69_all$phys/1000000

colnames(CML69_all)[2] <- "chr" 
#CML69_all  <- as.data.frame(lapply(CML69_all, gsub, pattern = "Chromosome ", replacement = ""))
CML69_all$chr <- sub("Chromosome ", "", CML69_all$chr)

CML69_all$chr <- as.integer(CML69_all$chr)


### combo plot
CML69_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(CML69_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(CML69_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =CML69_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="CML69") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
CML69_CO <- CML69_all
CML69_CO$chr <- as.factor(CML69_CO$chr)

CML69_CO_sum <- summarySE(CML69_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


CML69_COrate_bar <- ggplot(CML69_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("CML69F", "CML69M"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


CML69_nxo_sump <- ggplot(subset(COall_sum, line == "CML69"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
CML69F_nxo <- read.csv("../CML69/CML69F_numberCO.csv", row.names = 1 ) 

CML69F_nxo_df <- as.data.frame(CML69F_nxo) 
CML69F_nxo_df$set <- "CML69F"

CML69M_nxo <- read.csv("../CML69/CML69M_numberCO.csv", row.names = 1) 
CML69M_nxo_df <- as.data.frame(CML69M_nxo) 
CML69M_nxo_df$set <- "CML69M"

CML69_nxo <- rbind(CML69F_nxo_df, CML69M_nxo_df)

colnames(CML69_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


CML69_nxo_d <- melt(CML69_nxo, id = "set")
colnames(CML69_nxo_d) <- c("set", "chr", "CO")


CML69_nxo_sum <- summarySE(CML69_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


CML69_nxo_bar <- ggplot(CML69_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("CML69F", "CML69M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

CML69F_recomb_hot_genome <- read.csv("../CML69/hotspot/CML69F_recomb_hot_genome_0.2mb_5fold.csv" )
CML69M_recomb_hot_genome <- read.csv("../CML69/hotspot/CML69M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in CML69F_recomb_hot_genome$chr_Window
num_CML69F_recomb_hot <- length(CML69F_recomb_hot_genome$chr_Window)


# Get the number of rows in CML69M_recomb_hot_genome$chr_Window
num_CML69M_recomb_hot <- length(CML69M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_CML69_recomb_hot_genom <- intersect(CML69M_recomb_hot_genome$chr_Window, CML69F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_CML69_recomb_hot <- length(common_CML69_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package


CML69_hotspots <- venn.diagram(
  x = list(
    CML69F = CML69F_recomb_hot_genome$chr_Window,
    CML69M = CML69M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "CML69F" and "CML69M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(0.04, 0.04),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_CML69_recomb_hot  # Specify the size of the shared area
)


#grid.draw(CML69_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/CML69/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
CML69_phys_grouped <- CML69_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
CML69_phys_grouped2 <- CML69_phys_grouped[!is.na(CML69_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
CML69_mean_recomb_perwin <- CML69_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

CML69_mean_recomb_perwin_pos <- CML69_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
CML69_mean_recomb_perwinset <- CML69_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
CML69_mean_recomb_perwinset$CML69F[CML69_mean_recomb_perwinset$CML69F < 0] <- 0
CML69_mean_recomb_perwinset$CML69M[CML69_mean_recomb_perwinset$CML69M < 0] <- 0

# Genome average 
CML69_genome_recomb_rate <- mean(CML69_all$spline, na.rm = T )


#cutoff for general hotspots
CML69_recombhotspot_cutoff <- 5 * CML69_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
CML69_FM_hotspot <- CML69_mean_recomb_perwinset %>%
  filter((CML69F > 5 * CML69M | CML69M > 5 * CML69F) & 
         (CML69F > CML69_genome_recomb_rate/2 & CML69M > CML69_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(CML69F > CML69M, (CML69F + 0.00001) / (CML69M + 0.00001), -((CML69M + 0.00001) / (CML69F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(CML69F > CML69_recombhotspot_cutoff | CML69M > CML69_recombhotspot_cutoff, 
                             ifelse(CML69F > CML69M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(CML69_FM_hotspot, "CML69_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(CML69_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(CML69_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- CML69_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- CML69_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_CML69 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c(  "blue", "pink", "skyblue"), 
                    labels = c( "Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(CML69_recomb_combo, 
          ggarrange(CML69_nxo_bar, NA,
                    ggarrange(NA, CML69_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_CML69,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))


```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>

## Ki11
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

Ki11_all <- read.table("Ki11_mareyout.chr_snp.txt", sep = " " , header = TRUE )
Ki11_all$Mb <- Ki11_all$phys/1000000

colnames(Ki11_all)[2] <- "chr" 
#Ki11_all  <- as.data.frame(lapply(Ki11_all, gsub, pattern = "Chromosome ", replacement = ""))
Ki11_all$chr <- sub("Chromosome ", "", Ki11_all$chr)

Ki11_all$chr <- as.integer(Ki11_all$chr)


### combo plot
Ki11_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(Ki11_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(Ki11_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =Ki11_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="Ki11") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
Ki11_CO <- Ki11_all
Ki11_CO$chr <- as.factor(Ki11_CO$chr)

Ki11_CO_sum <- summarySE(Ki11_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


Ki11_COrate_bar <- ggplot(Ki11_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Ki11F", "Ki11M"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


Ki11_nxo_sump <- ggplot(subset(COall_sum, line == ""), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



##############################################################################

## CO number per chr
Ki11F_nxo <- read.csv("../Ki11/Ki11F_numberCO.csv", row.names = 1 ) 

Ki11F_nxo_df <- as.data.frame(Ki11F_nxo) 
Ki11F_nxo_df$set <- "Ki11F"

Ki11M_nxo <- read.csv("../Ki11/Ki11M_numberCO.csv", row.names = 1) 
Ki11M_nxo_df <- as.data.frame(Ki11M_nxo) 
Ki11M_nxo_df$set <- "Ki11M"

Ki11_nxo <- rbind(Ki11F_nxo_df, Ki11M_nxo_df)

colnames(Ki11_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


Ki11_nxo_d <- melt(Ki11_nxo, id = "set")
colnames(Ki11_nxo_d) <- c("set", "chr", "CO")


Ki11_nxo_sum <- summarySE(Ki11_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


Ki11_nxo_bar <- ggplot(Ki11_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Ki11F", "Ki11M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

Ki11F_recomb_hot_genome <- read.csv("../Ki11/hotspot/Ki11F_recomb_hot_genome_0.2mb_5fold.csv" )
Ki11M_recomb_hot_genome <- read.csv("../Ki11/hotspot/Ki11M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in Ki11F_recomb_hot_genome$chr_Window
num_Ki11F_recomb_hot <- length(Ki11F_recomb_hot_genome$chr_Window)


# Get the number of rows in Ki11M_recomb_hot_genome$chr_Window
num_Ki11M_recomb_hot <- length(Ki11M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_Ki11_recomb_hot_genom <- intersect(Ki11M_recomb_hot_genome$chr_Window, Ki11F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_Ki11_recomb_hot <- length(common_Ki11_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package


Ki11_hotspots <- venn.diagram(
  x = list(
    Ki11F = Ki11F_recomb_hot_genome$chr_Window,
    Ki11M = Ki11M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "Ki11F" and "Ki11M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(-0.3, -0.45),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_Ki11_recomb_hot  # Specify the size of the shared area
)


#grid.draw(Ki11_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/Ki11/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
Ki11_phys_grouped <- Ki11_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
Ki11_phys_grouped2 <- Ki11_phys_grouped[!is.na(Ki11_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
Ki11_mean_recomb_perwin <- Ki11_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

Ki11_mean_recomb_perwin_pos <- Ki11_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
Ki11_mean_recomb_perwinset <- Ki11_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
Ki11_mean_recomb_perwinset$Ki11F[Ki11_mean_recomb_perwinset$Ki11F < 0] <- 0
Ki11_mean_recomb_perwinset$Ki11M[Ki11_mean_recomb_perwinset$Ki11M < 0] <- 0

# Genome average 
Ki11_genome_recomb_rate <- mean(Ki11_all$spline, na.rm = T )


#cutoff for general hotspots
Ki11_recombhotspot_cutoff <- 5 * Ki11_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
Ki11_FM_hotspot <- Ki11_mean_recomb_perwinset %>%
  filter((Ki11F > 5 * Ki11M | Ki11M > 5 * Ki11F) & 
         (Ki11F > Ki11_genome_recomb_rate/2 & Ki11M > Ki11_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(Ki11F > Ki11M, (Ki11F + 0.00001) / (Ki11M + 0.00001), -((Ki11M + 0.00001) / (Ki11F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(Ki11F > Ki11_recombhotspot_cutoff | Ki11M > Ki11_recombhotspot_cutoff, 
                             ifelse(Ki11F > Ki11M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(Ki11_FM_hotspot, "Ki11_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(Ki11_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(Ki11_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- Ki11_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- Ki11_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_Ki11 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c( "blue", "pink", "skyblue"), 
                    labels = c( "Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(Ki11_recomb_combo, 
          ggarrange(Ki11_nxo_bar, NA,
                    ggarrange(NA, Ki11_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_Ki11,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))


```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>

## Mo17
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

Mo17_all <- read.table("Mo17_mareyout.chr_snp.txt", sep = " " , header = TRUE )
Mo17_all$Mb <- Mo17_all$phys/1000000

colnames(Mo17_all)[2] <- "chr" 
#Mo17_all  <- as.data.frame(lapply(Mo17_all, gsub, pattern = "Chromosome ", replacement = ""))
Mo17_all$chr <- sub("Chromosome ", "", Mo17_all$chr)

Mo17_all$chr <- as.integer(Mo17_all$chr)


### combo plot
Mo17_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(Mo17_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(Mo17_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =Mo17_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="Mo17") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
Mo17_CO <- Mo17_all
Mo17_CO$chr <- as.factor(Mo17_CO$chr)

Mo17_CO_sum <- summarySE(Mo17_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


Mo17_COrate_bar <- ggplot(Mo17_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Mo17F", "Mo17M"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


Mo17_nxo_sump <- ggplot(subset(COall_sum, line == "Mo17"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
Mo17F_nxo <- read.csv("../Mo17/Mo17F_numberCO.csv", row.names = 1 ) 

Mo17F_nxo_df <- as.data.frame(Mo17F_nxo) 
Mo17F_nxo_df$set <- "Mo17F"

Mo17M_nxo <- read.csv("../Mo17/Mo17M_numberCO.csv", row.names = 1) 
Mo17M_nxo_df <- as.data.frame(Mo17M_nxo) 
Mo17M_nxo_df$set <- "Mo17M"

Mo17_nxo <- rbind(Mo17F_nxo_df, Mo17M_nxo_df)

colnames(Mo17_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


Mo17_nxo_d <- melt(Mo17_nxo, id = "set")
colnames(Mo17_nxo_d) <- c("set", "chr", "CO")


Mo17_nxo_sum <- summarySE(Mo17_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


Mo17_nxo_bar <- ggplot(Mo17_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("Mo17F", "Mo17M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

Mo17F_recomb_hot_genome <- read.csv("../Mo17/hotspot/Mo17F_recomb_hot_genome_0.2mb_5fold.csv" )
Mo17M_recomb_hot_genome <- read.csv("../Mo17/hotspot/Mo17M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in Mo17F_recomb_hot_genome$chr_Window
num_Mo17F_recomb_hot <- length(Mo17F_recomb_hot_genome$chr_Window)


# Get the number of rows in Mo17M_recomb_hot_genome$chr_Window
num_Mo17M_recomb_hot <- length(Mo17M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_Mo17_recomb_hot_genom <- intersect(Mo17M_recomb_hot_genome$chr_Window, Mo17F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_Mo17_recomb_hot <- length(common_Mo17_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package


Mo17_hotspots <- venn.diagram(
  x = list(
    Mo17F = Mo17F_recomb_hot_genome$chr_Window,
    Mo17M = Mo17M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "Mo17F" and "Mo17M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(-0.43, -0.47),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_Mo17_recomb_hot  # Specify the size of the shared area
)


#grid.draw(Mo17_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/Mo17/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
Mo17_phys_grouped <- Mo17_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
Mo17_phys_grouped2 <- Mo17_phys_grouped[!is.na(Mo17_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
Mo17_mean_recomb_perwin <- Mo17_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

Mo17_mean_recomb_perwin_pos <- Mo17_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
Mo17_mean_recomb_perwinset <- Mo17_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
Mo17_mean_recomb_perwinset$Mo17F[Mo17_mean_recomb_perwinset$Mo17F < 0] <- 0
Mo17_mean_recomb_perwinset$Mo17M[Mo17_mean_recomb_perwinset$Mo17M < 0] <- 0

# Genome average 
Mo17_genome_recomb_rate <- mean(Mo17_all$spline, na.rm = T )


#cutoff for general hotspots
Mo17_recombhotspot_cutoff <- 5 * Mo17_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
Mo17_FM_hotspot <- Mo17_mean_recomb_perwinset %>%
  filter((Mo17F > 5 * Mo17M | Mo17M > 5 * Mo17F) & 
         (Mo17F > Mo17_genome_recomb_rate/2 & Mo17M > Mo17_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(Mo17F > Mo17M, (Mo17F + 0.00001) / (Mo17M + 0.00001), -((Mo17M + 0.00001) / (Mo17F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(Mo17F > Mo17_recombhotspot_cutoff | Mo17M > Mo17_recombhotspot_cutoff, 
                             ifelse(Mo17F > Mo17M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(Mo17_FM_hotspot, "Mo17_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(Mo17_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(Mo17_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- Mo17_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- Mo17_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_Mo17 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c( "red", "blue", "pink", "skyblue"), 
                    labels = c("Female higher (hotspot)", "Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(Mo17_recomb_combo, 
          ggarrange(Mo17_nxo_bar, NA,
                    ggarrange(NA, Mo17_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_Mo17,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))


```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>



## CML228
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

CML228_all <- read.table("CML228_mareyout.chr_snp.txt", sep = " " , header = TRUE )
CML228_all$Mb <- CML228_all$phys/1000000

colnames(CML228_all)[2] <- "chr" 
#CML228_all  <- as.data.frame(lapply(CML228_all, gsub, pattern = "Chromosome ", replacement = ""))
CML228_all$chr <- sub("Chromosome ", "", CML228_all$chr)

CML228_all$chr <- as.integer(CML228_all$chr)


### combo plot
CML228_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(CML228_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(CML228_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =CML228_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="CML228") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
CML228_CO <- CML228_all
CML228_CO$chr <- as.factor(CML228_CO$chr)

CML228_CO_sum <- summarySE(CML228_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


CML228_COrate_bar <- ggplot(CML228_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("CML228F", "CML228M"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


CML228_nxo_sump <- ggplot(subset(COall_sum, line == "CML228"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
CML228F_nxo <- read.csv("../CML228/CML228F_numberCO.csv", row.names = 1 ) 

CML228F_nxo_df <- as.data.frame(CML228F_nxo) 
CML228F_nxo_df$set <- "CML228F"

CML228M_nxo <- read.csv("../CML228/CML228M_numberCO.csv", row.names = 1) 
CML228M_nxo_df <- as.data.frame(CML228M_nxo) 
CML228M_nxo_df$set <- "CML228M"

CML228_nxo <- rbind(CML228F_nxo_df, CML228M_nxo_df)

colnames(CML228_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


CML228_nxo_d <- melt(CML228_nxo, id = "set")
colnames(CML228_nxo_d) <- c("set", "chr", "CO")


CML228_nxo_sum <- summarySE(CML228_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


CML228_nxo_bar <- ggplot(CML228_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("CML228F", "CML228M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

CML228F_recomb_hot_genome <- read.csv("../CML228/hotspot/CML228F_recomb_hot_genome_0.2mb_5fold.csv" )
CML228M_recomb_hot_genome <- read.csv("../CML228/hotspot/CML228M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in CML228F_recomb_hot_genome$chr_Window
num_CML228F_recomb_hot <- length(CML228F_recomb_hot_genome$chr_Window)


# Get the number of rows in CML228M_recomb_hot_genome$chr_Window
num_CML228M_recomb_hot <- length(CML228M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_CML228_recomb_hot_genom <- intersect(CML228M_recomb_hot_genome$chr_Window, CML228F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_CML228_recomb_hot <- length(common_CML228_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package


CML228_hotspots <- venn.diagram(
  x = list(
    CML228F = CML228F_recomb_hot_genome$chr_Window,
    CML228M = CML228M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "CML228F" and "CML228M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(-0.43, -0.47),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_CML228_recomb_hot  # Specify the size of the shared area
)


#grid.draw(CML228_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/CML228/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
CML228_phys_grouped <- CML228_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
CML228_phys_grouped2 <- CML228_phys_grouped[!is.na(CML228_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
CML228_mean_recomb_perwin <- CML228_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

CML228_mean_recomb_perwin_pos <- CML228_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
CML228_mean_recomb_perwinset <- CML228_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
CML228_mean_recomb_perwinset$CML228F[CML228_mean_recomb_perwinset$CML228F < 0] <- 0
CML228_mean_recomb_perwinset$CML228M[CML228_mean_recomb_perwinset$CML228M < 0] <- 0

# Genome average 
CML228_genome_recomb_rate <- mean(CML228_all$spline, na.rm = T )


#cutoff for general hotspots
CML228_recombhotspot_cutoff <- 5 * CML228_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
CML228_FM_hotspot <- CML228_mean_recomb_perwinset %>%
  filter((CML228F > 5 * CML228M | CML228M > 5 * CML228F) & 
         (CML228F > CML228_genome_recomb_rate/2 & CML228M > CML228_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(CML228F > CML228M, (CML228F + 0.00001) / (CML228M + 0.00001), -((CML228M + 0.00001) / (CML228F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(CML228F > CML228_recombhotspot_cutoff | CML228M > CML228_recombhotspot_cutoff, 
                             ifelse(CML228F > CML228M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(CML228_FM_hotspot, "CML228_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(CML228_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(CML228_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- CML228_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- CML228_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_CML228 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c( "red", "blue", "pink", "skyblue"), 
                    labels = c("Female higher (hotspot)", "Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(CML228_recomb_combo, 
          ggarrange(CML228_nxo_bar, NA,
                    ggarrange(NA, CML228_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_CML228,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))


```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>




## CML277
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

CML277_all <- read.table("CML277_mareyout.chr_snp.txt", sep = " " , header = TRUE )
CML277_all$Mb <- CML277_all$phys/1000000

colnames(CML277_all)[2] <- "chr" 
#CML277_all  <- as.data.frame(lapply(CML277_all, gsub, pattern = "Chromosome ", replacement = ""))
CML277_all$chr <- sub("Chromosome ", "", CML277_all$chr)

CML277_all$chr <- as.integer(CML277_all$chr)


### combo plot
CML277_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(CML277_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(CML277_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =CML277_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="CML277") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
CML277_CO <- CML277_all
CML277_CO$chr <- as.factor(CML277_CO$chr)

CML277_CO_sum <- summarySE(CML277_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


CML277_COrate_bar <- ggplot(CML277_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("CML277F", "CML277M"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


CML277_nxo_sump <- ggplot(subset(COall_sum, line == "CML277"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
CML277F_nxo <- read.csv("../CML277/CML277F_numberCO.csv", row.names = 1 ) 

CML277F_nxo_df <- as.data.frame(CML277F_nxo) 
CML277F_nxo_df$set <- "CML277F"

CML277M_nxo <- read.csv("../CML277/CML277M_numberCO.csv", row.names = 1) 
CML277M_nxo_df <- as.data.frame(CML277M_nxo) 
CML277M_nxo_df$set <- "CML277M"

CML277_nxo <- rbind(CML277F_nxo_df, CML277M_nxo_df)

colnames(CML277_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


CML277_nxo_d <- melt(CML277_nxo, id = "set")
colnames(CML277_nxo_d) <- c("set", "chr", "CO")


CML277_nxo_sum <- summarySE(CML277_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


CML277_nxo_bar <- ggplot(CML277_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("CML277F", "CML277M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

CML277F_recomb_hot_genome <- read.csv("../CML277/hotspot/CML277F_recomb_hot_genome_0.2mb_5fold.csv" )
CML277M_recomb_hot_genome <- read.csv("../CML277/hotspot/CML277M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in CML277F_recomb_hot_genome$chr_Window
num_CML277F_recomb_hot <- length(CML277F_recomb_hot_genome$chr_Window)


# Get the number of rows in CML277M_recomb_hot_genome$chr_Window
num_CML277M_recomb_hot <- length(CML277M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_CML277_recomb_hot_genom <- intersect(CML277M_recomb_hot_genome$chr_Window, CML277F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_CML277_recomb_hot <- length(common_CML277_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package


CML277_hotspots <- venn.diagram(
  x = list(
    CML277F = CML277F_recomb_hot_genome$chr_Window,
    CML277M = CML277M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "CML277F" and "CML277M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(-0.45, -0.45),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_CML277_recomb_hot  # Specify the size of the shared area
)


#grid.draw(CML277_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/CML277/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
CML277_phys_grouped <- CML277_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
CML277_phys_grouped2 <- CML277_phys_grouped[!is.na(CML277_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
CML277_mean_recomb_perwin <- CML277_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

CML277_mean_recomb_perwin_pos <- CML277_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
CML277_mean_recomb_perwinset <- CML277_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
CML277_mean_recomb_perwinset$CML277F[CML277_mean_recomb_perwinset$CML277F < 0] <- 0
CML277_mean_recomb_perwinset$CML277M[CML277_mean_recomb_perwinset$CML277M < 0] <- 0

# Genome average 
CML277_genome_recomb_rate <- mean(CML277_all$spline, na.rm = T )


#cutoff for general hotspots
CML277_recombhotspot_cutoff <- 5 * CML277_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
CML277_FM_hotspot <- CML277_mean_recomb_perwinset %>%
  filter((CML277F > 5 * CML277M | CML277M > 5 * CML277F) & 
         (CML277F > CML277_genome_recomb_rate/2 & CML277M > CML277_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(CML277F > CML277M, (CML277F + 0.00001) / (CML277M + 0.00001), -((CML277M + 0.00001) / (CML277F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(CML277F > CML277_recombhotspot_cutoff | CML277M > CML277_recombhotspot_cutoff, 
                             ifelse(CML277F > CML277M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(CML277_FM_hotspot, "CML277_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(CML277_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(CML277_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- CML277_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- CML277_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_CML277 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c( "blue", "pink", "skyblue"), 
                    labels = c("Male higher (hotspot)", "Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(CML277_recomb_combo, 
          ggarrange(CML277_nxo_bar, NA,
                    ggarrange(NA, CML277_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_CML277,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))



```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.
<div style="margin-bottom:150px;">
</div>



## CML103
<div style="margin-bottom:20px;">
</div>

```{r,  echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=12}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")
peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")

CML103_all <- read.table("CML103_mareyout.chr_snp.txt", sep = " " , header = TRUE )
CML103_all$Mb <- CML103_all$phys/1000000

colnames(CML103_all)[2] <- "chr" 
#CML103_all  <- as.data.frame(lapply(CML103_all, gsub, pattern = "Chromosome ", replacement = ""))
CML103_all$chr <- sub("Chromosome ", "", CML103_all$chr)

CML103_all$chr <- as.integer(CML103_all$chr)


### combo plot
CML103_recomb_combo <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
  geom_line(data = subset(CML103_all, spline >= 0), aes(x= Mb, y= spline*20, col = set), alpha =0.8, lwd =0.6) +
    #geom_area(data = subset(CML103_all, spline >= 0), aes(x= Mb, y= spline*20, color = set, fill = set, group = set), 
            # position = position_dodge(),alpha =0.4, size = 0.3) +
  geom_point(data =CML103_all, aes(x= Mb, y= gen, col = set), alpha =0.1, size = 0.5) +

  facet_wrap( ~ chr, ncol = 2, scales = "free") +
  theme_bw() +
  labs(x= "Physical positions (Mb)" ,
       y="Genetic distance (cM)", title="CML103") +
    scale_y_continuous(sec.axis = sec_axis(~ . / 20, name = "Recombination rates (cM/Mb)"))  +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 11),
        axis.title = element_text(size = 15),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 





# Recombination rate for each chromosome
CML103_CO <- CML103_all
CML103_CO$chr <- as.factor(CML103_CO$chr)

CML103_CO_sum <- summarySE(CML103_CO, measurevar = "spline", groupvars = c("set", "chr"), na.rm =T  )


CML103_COrate_bar <- ggplot(CML103_CO_sum, aes(x= set, y = spline , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20", alpha =0.8) +
  theme_bw() +
    labs(y="Recombination rate", title="", x="") +
    theme(axis.text.y=element_text(size=9), axis.text.x=element_blank(), axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=17), strip.text = element_text(size=12),
          legend.position = "right" , legend.title=element_blank(), legend.text=element_text(size=11)) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("CML103F", "CML103M"), labels = c("F", "M") )+
      geom_errorbar(aes(ymin=spline-se, ymax=spline+se),
                  size=.3, color="grey10",
                  width=.2,                    
                  position=position_dodge(.9)) +
   facet_wrap( ~ chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



# Number of crossover per individual


CML103_nxo_sump <- ggplot(subset(COall_sum, line == "CML103"), aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.8) +
  stat_summary(fun.data=mean_sd, geom="pointrange", color= "white", size = 0.6) +
  theme_classic() +
    labs(y="Number of crossover", title="", x="") +
  geom_signif(comparisons = list(c("F", "M")), 
              map_signif_level=TRUE) +
    theme(axis.text.y=element_text(size=9),axis.text.x=element_text(size=11),
          strip.background = element_blank(),
          axis.title=element_text(size=14), strip.text = element_text(size=12),
          legend.position = "none" ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


##############################################################################

## CO number per chr
CML103F_nxo <- read.csv("../CML103/CML103F_numberCO.csv", row.names = 1 ) 

CML103F_nxo_df <- as.data.frame(CML103F_nxo) 
CML103F_nxo_df$set <- "CML103F"

CML103M_nxo <- read.csv("../CML103/CML103M_numberCO.csv", row.names = 1) 
CML103M_nxo_df <- as.data.frame(CML103M_nxo) 
CML103M_nxo_df$set <- "CML103M"

CML103_nxo <- rbind(CML103F_nxo_df, CML103M_nxo_df)

colnames(CML103_nxo) <- c("1", "2", "3", "4", "5",
                        "6", "7", "8", "9", "10", "set")


CML103_nxo_d <- melt(CML103_nxo, id = "set")
colnames(CML103_nxo_d) <- c("set", "chr", "CO")


CML103_nxo_sum <- summarySE(CML103_nxo_d, measurevar = "CO", groupvars = c("set", "chr"), na.rm =T  )


CML103_nxo_bar <- ggplot(CML103_nxo_sum, aes(x= set, y = CO , fill = set)) +
  geom_bar(position=position_dodge(), stat="identity", color="grey20") +
  theme_bw() +
    labs(y="Number of crossover", title="", x="") +
    theme(axis.text.y=element_text(size=9),
          axis.text.x=element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank(),
          axis.title=element_text(size=16), strip.text = element_text(size=12),
          legend.position = "none" , legend.title=element_blank()) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("CML103F", "CML103M"),
                      labels= c("F", "M"))+
      geom_errorbar(aes(ymin=CO-se, ymax=CO+se),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  facet_wrap(~chr, ncol = 5, scales = "free_y") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



#########################
# Hotspots

CML103F_recomb_hot_genome <- read.csv("../CML103/hotspot/CML103F_recomb_hot_genome_0.2mb_5fold.csv" )
CML103M_recomb_hot_genome <- read.csv("../CML103/hotspot/CML103M_recomb_hot_genome_0.2mb_5fold.csv" )

# Get the number of rows in CML103F_recomb_hot_genome$chr_Window
num_CML103F_recomb_hot <- length(CML103F_recomb_hot_genome$chr_Window)


# Get the number of rows in CML103M_recomb_hot_genome$chr_Window
num_CML103M_recomb_hot <- length(CML103M_recomb_hot_genome$chr_Window)


# Find the common elements between the two vectors.
common_CML103_recomb_hot_genom <- intersect(CML103M_recomb_hot_genome$chr_Window, CML103F_recomb_hot_genome$chr_Window)

# Get the number of rows in common_elements
num_common_CML103_recomb_hot <- length(common_CML103_recomb_hot_genom)




# Assuming you have installed and loaded the VennDiagram package


CML103_hotspots <- venn.diagram(
  x = list(
    CML103F = CML103F_recomb_hot_genome$chr_Window,
    CML103M = CML103M_recomb_hot_genome$chr_Window
  ),
  category.names = c("F", "M"),  # Use "F" and "M" instead of "CML103F" and "CML103M"
  fill = c("red2", "blue2"),
  resolution = 500,   main = "Number of \nrecombination hotspots",   main.cex = 1.3,   fontfamily = "arial",   cat.fontfamily = "arial",   main.fontfamily = "arial",
  margin = 0.2,
  alpha = 0.5,
  cex = 1.5,
  cat.cex = 1.5,
  cat.pos = 0,      # Set to 0 to position category labels inside the circles
  cat.dist = c(0.045, 0.04),  # Increase the distance of category labels from the circles
  cat.col = c("red2", "blue2"),
  lty = "blank", filename = NULL,
  
  
  shared = num_common_CML103_recomb_hot  # Specify the size of the shared area
)


#grid.draw(CML103_hotspots)

###
setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/CML103/hotspot/")

# hotspot using recombination rates (adjust window size!) 

# Group by chromosome and window (0.2Mb)
CML103_phys_grouped <- CML103_all %>%
  dplyr::group_by(chr, set) %>%
  dplyr::mutate(window = ceiling(phys / 200000)) %>%
  dplyr::group_by(window, .add = TRUE)


# First, remove the NA values from the "spline" column.
CML103_phys_grouped2 <- CML103_phys_grouped[!is.na(CML103_phys_grouped$spline), ]

# Group by "chr" and "window", and calculate the average spline per group.
CML103_mean_recomb_perwin <- CML103_phys_grouped2 %>%
  dplyr::group_by(set, chr, window) %>%
  dplyr::summarize(mean_recomb = mean(spline))


# Calculate the start and end positions of the windows based on the "window" column.
window_size_mb <- 0.2  # Define the window size in Mb

CML103_mean_recomb_perwin_pos <- CML103_mean_recomb_perwin %>%
  mutate(
    start = (window - 1) * window_size_mb,
    end = window * window_size_mb
  )

#Split into two columns for two sets
CML103_mean_recomb_perwinset <- CML103_mean_recomb_perwin_pos %>%
  pivot_wider(names_from = set, values_from = mean_recomb)

#Convert negative values to 0
CML103_mean_recomb_perwinset$CML103F[CML103_mean_recomb_perwinset$CML103F < 0] <- 0
CML103_mean_recomb_perwinset$CML103M[CML103_mean_recomb_perwinset$CML103M < 0] <- 0

# Genome average 
CML103_genome_recomb_rate <- mean(CML103_all$spline, na.rm = T )


#cutoff for general hotspots
CML103_recombhotspot_cutoff <- 5 * CML103_genome_recomb_rate  #5-fold higher than genome average



# Extract FM contrast pots (5 fold, higher than 1/2 genome average, whether it is general hotspots)
CML103_FM_hotspot <- CML103_mean_recomb_perwinset %>%
  filter((CML103F > 5 * CML103M | CML103M > 5 * CML103F) & 
         (CML103F > CML103_genome_recomb_rate/2 & CML103M > CML103_genome_recomb_rate/2)) %>%
  mutate(
    FM_ratio = ifelse(CML103F > CML103M, (CML103F + 0.00001) / (CML103M + 0.00001), -((CML103M + 0.00001) / (CML103F + 0.00001))),
    #if F>M, the ratio is positive but F<M it will be negative.
    general_hotspot = ifelse(CML103F > CML103_recombhotspot_cutoff | CML103M > CML103_recombhotspot_cutoff, 
                             ifelse(CML103F > CML103M, "F", "M"), "")
  )

# Sava the FM hotspot results
#write.csv(CML103_FM_hotspot, "CML103_FM_hotspot.csv", row.names = F )


# Count rows with positive FM_ratio
Count_FoverM <- sum(CML103_FM_hotspot$FM_ratio > 0)


# Count rows with negative FM_ratio
Count_MoverF <- sum(CML103_FM_hotspot$FM_ratio < 0)



# Create a vector to store the counts
counts <- c(positive = Count_FoverM, negative = Count_MoverF)


# Filter the data for Female higher and Male higher separately
female_higher <- CML103_FM_hotspot %>%
  filter(FM_ratio > 0)

male_higher <- CML103_FM_hotspot %>%
  filter(FM_ratio < 0)

# Create a function to calculate the counts for each category in general_hotspot
count_categories <- function(data, group_name) {
  data %>%
    dplyr::group_by(general_hotspot) %>%
    dplyr::summarise(count = n()) %>%
    dplyr::mutate(general_hotspot = ifelse(general_hotspot == "", "no general hotspot", general_hotspot),
           group = group_name,
           percent = count / sum(count))
}

# Calculate counts for Female higher and Male higher
female_counts <- count_categories(female_higher, "Female higher")
male_counts <- count_categories(male_higher, "Male higher")

# Combine the data and create a new category for "no general hotspot" within each group
combined_counts <- rbind(
  female_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Female higher)", sep = "")),
  male_counts %>% mutate(general_hotspot_grouped = paste(general_hotspot, " (Male higher)", sep = ""))
)



# Create the stacked bar plot using ggplot2
FM_hot_CML103 <- ggplot(combined_counts, aes(x = group, y = count, fill = general_hotspot_grouped)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of F vs M\ncontrast spots",
    fill = "General Hotspot",
    title = ""
  ) +
  theme_classic() +
  scale_fill_manual(values = c( "pink", "skyblue"), 
                    labels = c("Female higher", "Male higher")) +
    theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_text(size = 13),
    axis.title.x = element_blank(),
    legend.position = "bottom",
    legend.title = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 4)) 




# Merge plots
ggarrange(CML103_recomb_combo, 
          ggarrange(CML103_nxo_bar, NA,
                    ggarrange(NA, CML103_hotspots, NA, nrow =3, heights = c(0.45,1,0.45)), NA,
                    FM_hot_CML103,ncol =5, widths = c(1.3,0.1, 0.6, 0.1, 0.76)), 
          nrow = 2, heights = c(1,0.6))


```
<div style="margin-bottom:5px;">
</div>
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average.
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.

<div style="margin-bottom:150px;">
</div>


# 2. Recombination differences across lines
<div style="margin-bottom:50px;">
</div>


## 1) CO numbers of all lines

<div style="margin-bottom:20px;">
</div>

```{r, echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=4}


setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage")


csv_files <- dir(pattern='*numberCO.csv$', recursive = T)


for(i in 1:length(csv_files)) {
  if(i == 1)
    COall <- read.csv(csv_files[i])
  else
    COall <- rbind(COall, read.csv(csv_files[i]))
}

colnames(COall)[1] <- c("set")

COall$sum <- rowSums(COall[,2:11]) 

COall_sum <- COall[,c(1,12)]
colnames(COall_sum) <- c("set", "CO")

COall_sum$line <- sub("F_", "_", COall_sum$set)
COall_sum$line <- sub("M_", "_", COall_sum$line)
COall_sum$line <- sub("_.*", "", COall_sum$line)
COall_sum$line <- sub("_.*", "", COall_sum$line)
COall_sum$sex <- str_sub(COall_sum$set,-5,-5)

COall_sum <- COall_sum[c("line", "sex", "set", "CO")] 
COall_sum$line <- factor(COall_sum$line, levels = c("B97", "Oh43", "Il14H", "W22", "Mo18W","Oh7B","A344", "M162W", 
                                                    "CML322", "CML69", "Ki11", "CML228", "Mo17",  "CML277", "CML103"))
##write.csv(COall_sum, "COnumber_sum_all.csv", row.names = F)




########################
# Extract genetic distance (cM) per chr per set

cM_per_set_chr <- B97_all %>%
  dplyr::group_by(set, chr) %>%
  dplyr::summarise(max_gen = max(gen))

cM_per_Set <-  cM_per_set_chr %>%
  dplyr::group_by(set) %>%
  dplyr::summarise(sum_cM = sum(max_gen))



########################



#mean +-sd
COall_sump <- ggplot(COall_sum, aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.7, color = "gray50", linewidth =0.3) +
 stat_summary(fun = mean,
               geom = "pointrange",
               fun.max = function(x) mean(x) + sd(x),
               fun.min = function(x) mean(x) - sd(x), 
              color= "white", size= 0.5, linewidth = 0.5) +
  theme_classic() +
    labs(y="Number of crossover", title="Wilcox", x="BC1s crossed with B73") +
   facet_wrap( ~ line, nrow = 1, strip.position = "bottom") +
  geom_signif(comparisons = list(c("F", "M")),  map_signif_level=TRUE, tip_length = .01, test = "wilcox.test") +
    theme(axis.text.y=element_text(size=10), axis.text.x=element_blank(), axis.ticks.x=element_blank(), 
          strip.background = element_blank(), strip.placement = "outside", 
          axis.title=element_text(size=14), strip.text = element_text(size=10, margin = margin(0,0,20,0)), 
          legend.position = "bottom",  legend.justification = "right",
          legend.title=element_blank(), legend.margin = margin(-27, 0, 0, 0),
          plot.title = element_blank() ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M")) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


COall_sump


#t_test

COall_sumpt <- ggplot(COall_sum, aes(x= sex, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.7, color = "gray50", linewidth =0.3) +
 stat_summary(fun = mean,
               geom = "pointrange",
               fun.max = function(x) mean(x) + sd(x),
               fun.min = function(x) mean(x) - sd(x), 
              color= "white", size= 0.5, linewidth = 0.5) +
  theme_classic() + 
    labs(y="Number of crossover", title="Wilcox", x="BC1s crossed with B73") +
   facet_wrap( ~ line, nrow = 1, strip.position = "bottom") +
  geom_signif(comparisons = list(c("F", "M")),  map_signif_level=TRUE, tip_length = .01, test = "t.test") +
    theme(axis.text.y=element_text(size=10), axis.text.x=element_blank(), axis.ticks.x=element_blank(), 
          strip.background = element_blank(), strip.placement = "outside", 
          axis.title=element_text(size=14), strip.text = element_text(size=10, margin = margin(0,0,20,0)), 
          legend.position = "bottom",  legend.justification = "right",
          legend.title=element_blank(), legend.margin = margin(-27, 0, 0, 0),
          plot.title = element_blank() ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M")) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


COall_sumpt

###

# CO numbers in ch

COall$Line  <- sub("_.*", "", COall$set)
COall$sex  <- substr(COall$Line, nchar(COall$Line), nchar(COall$Line))
COall$Line <- substr(COall$Line , 1, nchar(COall$Line ) - 1)




##






###
# Reorder levels for Line variable
COall$Line <- factor(COall$Line, levels = c("B97", "Oh43", "Il14H", "W22", "Mo18W", "Oh7B", 
                                            "A344", "M162W", "CML322", "CML69", "Ki11", "CML228",
                                            "Mo17", "CML277", "CML103"))

#chr4

#stat for chr4

# Create an empty list to store the results
wilcox_results_4 <- list()

# Loop through unique lines and perform Wilcoxon test
unique_lines_4 <- unique(COall$Line)
for (line in unique_lines_4) {
  subset_data_4 <- subset(COall, Line == line)
  wilcox_test_4 <- wilcox.test(subset_data_4$X4 ~ subset_data_4$sex)
  wilcox_results_4[[line]] <- wilcox_test_4$p.value
}

# Combine p-values with Line names
wilcox_results_df_4 <- data.frame(Line = names(wilcox_results_4), p_value = unlist(wilcox_results_4))



COall_stat_X4 <- summarySE(COall, measurevar = "X4", groupvars = c("Line", "sex"))

CHR4 <- ggplot(data = COall_stat_X4, aes(x = Line, y= X4, fill = sex)) +
  geom_bar(position="dodge", stat="identity", color = "grey20") +
  theme_bw() +
  scale_fill_manual(values = c("red1", "blue4", "white")) +
  labs(y = "Number of COs", title = "Chromosome 4", x = "") +
  geom_errorbar(aes(ymin=X4-se, ymax=X4+se),
                  size=.3, color = "grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  theme(axis.text.y = element_text(size = 10), axis.text.x = element_blank(), 
        axis.title.y = element_text(size = 14), axis.title.x = element_blank(),
        legend.position = "none", 
        panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
      geom_text(aes(x=Line, y=X4+se+0.1, 
                  label=c( "","***",  "","**","","*","", "*","", "*", "","ns","","ns","","ns",
                           "","ns", "","ns","", "ns","", "ns","","ns","", "ns", "","***")),
                  position=position_dodge(width=0.9), size=4)

###


#chr6
COall_stat_X6 <- summarySE(COall, measurevar = "X6", groupvars = c("Line", "sex"))

CHR6 <- ggplot(data = COall_stat_X6, aes(x = Line, y= X6, fill = sex)) +
  geom_bar(position="dodge", stat="identity", color = "grey20") +
  theme_bw() +
  scale_fill_manual(values = c("red1", "blue4", "white")) +
  labs(y = "Number of COs", title = "Chromosome 6", x = "") +
  geom_errorbar(aes(ymin=X6-se, ymax=X6+se),
                  size=.3, color = "grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  theme(axis.text.y = element_text(size = 10), axis.text.x = element_blank(), 
        axis.title.y = element_text(size = 14), axis.title.x = element_blank(),
        legend.position = "none", 
        panel.grid.major = element_blank(), panel.grid.minor = element_blank()) 


#chr2
COall_stat_X2 <- summarySE(COall, measurevar = "X2", groupvars = c("Line", "sex"))

CHR2 <- ggplot(data = COall_stat_X2, aes(x = Line, y= X2, fill = sex)) +
  geom_bar(position="dodge", stat="identity", color = "grey20") +
  theme_bw() +
  scale_fill_manual(values = c("red1", "blue4", "white")) +
  labs(y = "Number of COs", title = "Chromosome 2", x = "BC1s crossed with B73") +
  geom_errorbar(aes(ymin=X2-se, ymax=X2+se),
                  size=.3, color = "grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  theme(axis.text.y = element_text(size = 10), axis.text.x = element_text(size = 10), 
        axis.title = element_text(size = 14),
        legend.position = "none", 
        panel.grid.major = element_blank(), panel.grid.minor = element_blank()) 

#chr7

#stat for chr4

# Create an empty list to store the results
wilcox_results_7 <- list()

# Loop through unique lines and perform Wilcoxon test
unique_lines_7 <- unique(COall$Line)
for (line in unique_lines_7) {
  subset_data_7 <- subset(COall, Line == line)
  wilcox_test_7 <- wilcox.test(subset_data_7$X7 ~ subset_data_7$sex)
  wilcox_results_7[[line]] <- wilcox_test_7$p.value
}

# Combine p-values with Line names
wilcox_results_df_7 <- data.frame(Line = names(wilcox_results_7), p_value = unlist(wilcox_results_7))




COall_stat_X7 <- summarySE(COall, measurevar = "X7", groupvars = c("Line", "sex"))

CHR7 <- ggplot(data = COall_stat_X7, aes(x = Line, y= X7, fill = sex)) +
  geom_bar(position="dodge", stat="identity", color = "grey20") +
  theme_bw() +
  scale_fill_manual(values = c("red1", "blue4", "white")) +
  labs(y = "Number of COs", title = "Chromosome 7", x = "BC1s crossed with B73") +
  geom_errorbar(aes(ymin=X7-se, ymax=X7+se),
                  size=.3, color = "grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  theme(axis.text.y = element_text(size = 10), axis.text.x = element_text(size = 10), 
        axis.title = element_text(size = 14),
        legend.position = "none", 
        panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
        geom_text(aes(x=Line, y=X7+se+0.1, 
                  label=c( "","*",  "","**","","ns","", "ns","", "ns", "","ns","","ns","","ns",
                           "","ns", "","ns","", "ns","", "ns","","ns", "", "ns", "","**")),
                  position=position_dodge(width=0.9), size=4)




merged_CHRCO4 <- ggarrange(CHR4,  CHR6,  CHR2, CHR7, 
          ncol = 2, nrow = 2, align = "hv")

merged_CHRCO2 <- ggarrange(CHR4,  CHR7, 
          nrow = 2, align = "hv" )

#png(filename = "plot/COinchr4and7.png", width = 9, height = 5.5, units = "in",  res = 400 )
merged_CHRCO2
#dev.off()



#Mean_CO for tropical and non-tropical

mean_co <- mean_co %>%
  dplyr::mutate(
    Origin = case_when(
      line %in% c("B97", "Oh43", "Il14H", "W22", "Oh7B", "A344", "Mo17", "Mo18W", "M162W") ~ "Non-tropical",
      line %in% c("CML322", "CML69", "Ki11", "CML228", "CML277", "CML103") ~ "Tropical",
      TRUE ~ NA_character_  # For lines not in either list
    )
  )



mean_co$OriginSex <- with(mean_co, paste(Origin, sex, sep = "_"))


mean_co.aov <- aov( mean_CO ~ Origin*sex, data = mean_co)
#shapiro.test(mean_co.aov$residuals) #Normality?
#TukeyHSD(mean_co.aov)  

mean_co_res <- HSD.test(mean_co.aov, c("Origin", "sex"), group=TRUE)
#mean_co_res$groups

oder.mean_co_res <- mean_co_res$groups[order(rownames(mean_co_res$groups)),] 



Mean_CO_Trop <- summarySE(mean_co, measurevar="mean_CO", groupvars=c("Origin","sex"))


################
#Bar plot


Mean_CO_Trop_p <- ggplot(Mean_CO_Trop, aes(x = Origin, y = mean_CO, fill = sex)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_classic() +
   #  facet_wrap( ~ Origin, nrow = 1, strip.position = "bottom", scales = "free_x" ) +
  labs(y = "Number of crossover", title = "Wilcox", x = "") +
  theme(axis.text.y = element_text(size = 12), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        strip.background = element_blank(), 
        strip.placement = "outside", 
        axis.title = element_text(size = 15), 
        strip.text = element_text(size = 14, margin = margin(0, 0, 20, 0)), 
        legend.position = "right", 
        legend.justification = "right",
        legend.title = element_blank(), 
        plot.title = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
  scale_fill_manual(values = c("red2", "blue4"),
                    breaks = c("F", "M")) +
  geom_errorbar(aes(ymin=mean_CO-sd, ymax=mean_CO+sd),
                  size=.3, color="grey20",
                  width=.2,                    
                  position=position_dodge(.9)) +
  geom_text(aes(x=Origin, y=mean_CO+sd+0.8, 
                  label=c(as.character(oder.mean_co_res$groups))),
             position=position_dodge(width=0.9), size=5) 




#box
Mean_CO_Trop_p2 <- ggplot(mean_co, aes(x = Origin, y = mean_CO, fill = sex)) +
  geom_boxplot(outlier.shape = NA, width = 0.9, color = "grey40") +
  theme_classic() +
  facet_wrap(~ Origin, nrow = 1, strip.position = "bottom", scales = "free_x") +
  labs(y = "Number of crossover", title = "", x = "Origin") +
  theme(axis.text.y = element_text(size = 12), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        strip.background = element_blank(), 
        strip.placement = "outside", 
        axis.title = element_text(size = 15), 
        strip.text = element_text(size = 13, margin = margin(0, 0,5, 0)), 
        legend.position = c(0.93, 0.01),
        legend.justification = "bottom",  
        legend.title = element_blank(), 
        plot.title = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
  scale_fill_manual(values = c("red2", "blue4"),
                    breaks = c("F", "M")) +
  geom_text(data= Mean_CO_Trop, aes(x=Origin, y=c(13.3,14.25,13.8,13.3), 
                  label=c(as.character(oder.mean_co_res$groups))),
             position=position_dodge(width=0.9), size=5)

Mean_CO_Trop_p2


#png(filename = "plot/Tro_Ntro.png", width = 3.5, height = 4, units = "in",  res = 400 )
#Mean_CO_Trop_p2
#dev.off()


```
<a href="#top">Back to top</a>

<div style="margin-bottom:150px;">
</div>


## 2) Variation of CO numbers in females and males, respectively
<div style="margin-bottom:20px;">
</div>
```{r, echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=4}

# Your data and code for COall.aov and COall_res here...

COall_sum$line <- factor(COall_sum$line, levels = c("B97", "Oh43", "Il14H", "W22", "Oh7B","A344", "CML277","M162W", "Mo18W",
                                                    "CML322", "CML69", "Ki11", "CML228", "Mo17",  "CML103"))

COall_sum_sex <- ggplot(COall_sum, aes(x = line, y = CO, fill = sex)) +
  geom_violin(trim = TRUE, bw = 1, alpha = 0.7) +
  stat_summary(
    fun = mean,
    geom = "pointrange",
    fun.max = function(x) mean(x) + sd(x),
    fun.min = function(x) mean(x) - sd(x),
    color = "white",
    size = 0.5,
    linewidth = 0.5
  ) +
  theme_classic() +
  labs(y = "Number of crossover", title = "", x = "BC1s crossed with B73") +
  facet_wrap(sex ~ ., ncol = 2, strip.position = "bottom") +  # Separate facets for each sex
  theme(
    axis.text.y = element_text(size = 10),
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels for better visibility
    strip.background = element_blank(),
    strip.placement = "outside",
    axis.title = element_text(size = 14),
    strip.text = element_blank(),
    legend.position = "bottom",
    legend.justification = "right",
    legend.title = element_blank(),
    legend.margin = margin(-27, 0, 0, 0),
    plot.title = element_blank()
  ) +
  scale_fill_manual(values = c("red2", "blue4"), breaks = c("F", "M")) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

COall_sum_sex


  
```
<a href="#top">Back to top</a>

<div style="margin-bottom:150px;">
</div>



## 3) Total genetic distance (cM) of all lines
<div style="margin-bottom:20px;">
</div>
```{r, echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=3}

# List of data frame names
df_names <- c("B97_all", "Oh43_all", "Il14H_all", "W22_all", "Mo18W_all", "Oh7B_all",
              "A344_all", "M162W_all", "CML322_all", "CML69_all", "Ki11_all",
              "CML228_all", "Mo17_all", "CML277_all", "CML103_all")

# Function to extract genetic distance (cM) per chr per set and sum the max_gen per set
extract_and_sum_cM <- function(df_name) {
  df <- get(df_name)  # Get the data frame based on its name
  cM_per_set_chr <- dplyr::group_by(df, set, chr) %>%
    dplyr::summarise(max_gen = max(gen))
  
  cM_per_Set <- dplyr::group_by(cM_per_set_chr, set) %>%
    dplyr::summarise(sum_cM = sum(max_gen))
  
  # Separate the 'set' column into 'line' and 'sex'
  cM_per_Set <- tidyr::separate(cM_per_Set, set, into = c("line", "sex"), sep = -1, remove = FALSE)
  
  return(cM_per_Set)
}

# Loop through the data frame names and apply the function to each data frame
cM_per_Set_list <- lapply(df_names, extract_and_sum_cM)

# Merge the list of data frames into a single data frame
merged_cM_per_Set <- dplyr::bind_rows(cM_per_Set_list)


# Define the desired order of line names
line_order <- c("B97", "Oh43", "Il14H", "W22", "Mo18W", "Oh7B", "A344", "M162W", 
                "CML322", "CML69", "Ki11", "CML228", "Mo17",  "CML277", "CML103")

# Set the "line" variable as a factor with the desired order
merged_cM_per_Set$line <- factor(merged_cM_per_Set$line, levels = line_order)

# Plot sum_cM for F and M for each line using ggplot2
ggplot(merged_cM_per_Set, aes(x = line, y = sum_cM, fill = sex)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "BC1s crossed with B73", y = "Genetic distance across \nall chromosomes (cM)") +
  scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M")) +
  theme_classic() +
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.ticks.x=element_blank(), 
          strip.background = element_blank(), strip.placement = "outside", 
          axis.title=element_text(size=14), strip.text = element_text(size=10, margin = margin(0,0,20,0)), 
          legend.position = "bottom",  legend.justification = "right",
          legend.title=element_blank(), legend.margin = margin(-27, 0, 0, 0),
          plot.title = element_blank() )  +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())



##write.csv(merged_cM_per_Set, "../Total_cM_all_lines.csv", row.names =F )


```

<a href="#top">Back to top</a>

<div style="margin-bottom:150px;">
</div>





## 4) Correlation between genetic distance (crossover numbers) and methylation 
##### *Methylation data was obtained from Huffold et al., 2021

<div style="margin-bottom:20px;">
</div>

```{r, echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=5.5}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/Methylation")
methyl_cm <- read.csv("Methylation_cM.csv")

methyl_cm$Line <- factor(methyl_cm$Line, levels = c("B97", "Oh43", "Il14H",  "Mo18W","Oh7B", "M162W", 
                                                     "CML69", "Ki11", "CML228", "CML277", "CML103"))

# Create a custom blue to red color palette
blue_to_red_palette <- colorRampPalette(colors = c("blue", "red"))

# Generate 10 colors from the custom palette
my_palette <- blue_to_red_palette(11)


CHH_Female <- ggscatter(methyl_cm, x = "CHH", y = "Female_cM", 
                       color ="Line",
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          main = "CHH",
          add.params = list(color = "black", fill = "grey80"),
          xlab = "", ylab = "") + 
          theme_bw() +
          theme(panel.grid.major = element_blank()) +
          theme(panel.grid.minor = element_blank())+
             theme(plot.title = element_text(hjust = 0.5),  legend.text = element_text(size = 12) ) +
          scale_color_manual(values = my_palette) +
          scale_x_continuous(breaks = c(0.012, 0.013, 0.014)) 

CHH_Female_co <- ggscatter(methyl_cm, x = "CHH", y = "Female_CO", 
                       color ="Line",
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          main = "CHH",
          add.params = list(color = "black", fill = "grey80"),
          xlab = "", ylab = "") + 
          theme_bw() +
          theme(panel.grid.major = element_blank()) +
          theme(panel.grid.minor = element_blank())+
             theme(plot.title = element_text(hjust = 0.5),  legend.text = element_text(size = 12) ) +
          scale_color_manual(values = my_palette) +
          scale_x_continuous(breaks = c(0.012, 0.013, 0.014)) #lower correlation when using COs instead of cM!

CHH_Male <- ggscatter(methyl_cm, x = "CHH", y = "Male_cM", 
                       color ="Line",
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          main = "",
          add.params = list(color = "black", fill = "grey80"),
          xlab = "", ylab = "") +
          theme_bw() +
          theme(panel.grid.major = element_blank()) +
          theme(panel.grid.minor = element_blank())+
             theme(plot.title = element_text(hjust = 0.5), legend.text = element_text(size = 12))+
          scale_color_manual(values = my_palette)+
          scale_x_continuous(breaks = c(0.012, 0.013, 0.014)) 



CG_Female <- ggscatter(methyl_cm, x = "CG", y = "Female_cM", 
                       color ="Line",
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          main = "CG",
          add.params = list(color = "black", fill = "grey80"),
          xlab = "", ylab = "Genetic distance (cM) \nin females") +
          theme_bw() +
          theme(panel.grid.major = element_blank()) +
          theme(panel.grid.minor = element_blank())+
             theme(plot.title = element_text(hjust = 0.5), legend.text = element_text(size = 12))+
          scale_color_manual(values = my_palette)


CG_Male <- ggscatter(methyl_cm, x = "CG", y = "Male_cM", 
                       color ="Line",
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          main = "",
          add.params = list(color = "black", fill = "grey80"),
          xlab = "", ylab = "Genetic distance (cM) \nin males") +
          theme_bw() +
          theme(panel.grid.major = element_blank()) +
          theme(panel.grid.minor = element_blank())+
             theme(plot.title = element_text(hjust = 0.5), legend.text = element_text(size = 12))+
          scale_color_manual(values = my_palette)


CHG_Female <- ggscatter(methyl_cm, x = "CHG", y = "Female_cM", 
                       color ="Line",
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          main = "CHG",
          add.params = list(color = "black", fill = "grey80"),
          xlab = "", ylab = "") +
          theme_bw() +
          theme(panel.grid.major = element_blank()) +
          theme(panel.grid.minor = element_blank())+
             theme(plot.title = element_text(hjust = 0.5), legend.text = element_text(size = 12))+
          scale_color_manual(values = my_palette)

  
CHG_Male <- ggscatter(methyl_cm, x = "CHG", y = "Male_cM", 
                       color ="Line",
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          main = "",
          add.params = list(color = "black", fill = "grey80"),
          xlab = "", ylab = "") +
          theme_bw() +
          theme(panel.grid.major = element_blank()) +
          theme(panel.grid.minor = element_blank())+
             theme(plot.title = element_text(hjust = 0.5), legend.text = element_text(size = 12))+
          scale_color_manual(values = my_palette)



# Merge


merged_cor <- ggarrange(CG_Female,  CHG_Female,  CHH_Female, CG_Male, CHG_Male, CHH_Male, 
          ncol = 3, nrow = 2, align = "hv", common.legend = T, legend = "right" )

merged_cor2 <- annotate_figure(merged_cor,
                bottom = text_grob("Methylation     ", color = "black",  size = 13))



#png(filename = "plot/methylation_cM_all.png", width = 10, height = 5.5, units = "in",  res = 400 )
merged_cor2
#dev.off()


```


<a href="#top">Back to top</a>

<div style="margin-bottom:150px;">
</div>

```{r, include=F, echo=FALSE, message=FALSE, warning=FALSE}
# Correlation between CO numbers and methylation

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/")

mean_COnum <- COall_sum %>%
  dplyr::group_by(line, sex) %>%
  dplyr::summarise(mean_CO = mean(CO, na.rm = TRUE))

##write.csv(mean_COnum, "Mean_COperline.csv", row.names = F)

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/Methylation")
methyl_cm <- read.csv("Methylation_cM.csv")

methyl_cm$Line <- factor(methyl_cm$Line, levels = c("B97", "Oh43", "Il14H",  "Mo18W","Oh7B", "M162W", 
                                                     "CML69", "Ki11",  "CML277", "CML103"))



CHH_FemaleCO <- ggscatter(methyl_cm, x = "CHH", y = "Female_CO", 
                       color ="Line",
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          main = "CHH",
          add.params = list(color = "black", fill = "grey80"),
          xlab = "Methylation", ylab = "Genetic distance (cM) in females") +
          theme_bw() +
          theme(panel.grid.major = element_blank()) +
          theme(panel.grid.minor = element_blank())+
             theme(plot.title = element_text(hjust = 0.5), legend.text = element_text(size = 12))+
          scale_color_manual(values = my_palette)

CHH_FemaleCO



```








## 5) Recombination hotspots for each chromosome
##### *Recombination hotspots were identified in 0.2Mb windows with at least 5-fold higher than the genome average. 


<div style="margin-bottom:20px;">
</div>
```{r, echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=12, fig.height=10}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage")


hotspot_files <- dir(pattern = '*_recomb_hot_genome_0.2mb_5fold.csv$', recursive = TRUE)

# Create an empty list to store the data frames
hotspotall_list <- list()

# Read the files and store them in the list along with the corresponding sample name
for (i in 1:length(hotspot_files)) {
  df <- read.csv(hotspot_files[i])
  # Extract the sample name from the file name
  sample_name <- sub("_recomb_hot_genome_0\\.2mb_5fold\\.csv$", "", basename(hotspot_files[i]))
  # Extract the last character as sex
  sex <- substr(sample_name, nchar(sample_name), nchar(sample_name))
  # Extract all characters except the last one as line
  line <- substr(sample_name, 1, nchar(sample_name) - 1)
  
  # Add new columns with the sample information
  df$Sex <- sex
  df$Line <- line
  
  # Add the hotspot_no column based on the Sex column
  df$hotspot_no <- ifelse(sex == "F", 1, ifelse(sex == "M", 2, NA))
  
  # Store the data frame in the list
  hotspotall_list[[i]] <- df
}

# Combine all the data frames into a single data frame
hotspotall <- do.call(rbind, hotspotall_list)


# Split data into different sexes
hotspot_females <- subset(hotspotall, Sex == "F")[,c(4,6,7)]
hotspot_males <- subset(hotspotall, Sex == "M")[,c(4,6,7)]


# Merge data based on chr_Window and Line
merged_data <- full_join(hotspot_females, hotspot_males, by = c("chr_Window", "Line"), suffix = c("_F", "_M"))

# Calculate the final hotspot_no based on the presence of F and M per chr_Window and Line
merged_data <- merged_data %>%
  dplyr::mutate(hotspot_no = case_when(
    !is.na(hotspot_no_F) & is.na(hotspot_no_M) ~ hotspot_no_F,
    is.na(hotspot_no_F) & !is.na(hotspot_no_M) ~ hotspot_no_M,
    !is.na(hotspot_no_F) & !is.na(hotspot_no_M) ~ 3
  )) %>%
  select(Line, chr_Window, hotspot_no)

# Remove duplicates if any (to ensure only one row per chr_Window and Line combination)
merged_data <- distinct(merged_data)

# Order the result by Line and chr_Window
merged_data <- merged_data %>%
  arrange(Line, chr_Window)



##################


# Define the desired window ranges for each chromosome
chromosomes <- 1:10
window_ranges <- list(
  c(1:1543),
  c(1:1219),
  c(1:1142),
  c(1:1252),
  c(1:1132),
  c(1:907),
  c(1:930),
  c(1:913),
  c(1:816),
  c(1:763)
)


# Create a data frame with all chr_Window combinations based on the desired window ranges
chr_window_data <- data.frame(chr_Window = character(),
                              stringsAsFactors = FALSE)

for (chrom in chromosomes) {
  window_range <- window_ranges[[chrom]]
  chr_Window <- paste0(chrom, "_", window_range)
  chr_window_data <- rbind(chr_window_data, data.frame(chr_Window, stringsAsFactors = FALSE))
}

# Create a data frame to store the final merged data
final_data <- data.frame(chr_window_data, stringsAsFactors = FALSE)

# Get unique lines from the merged_data
unique_lines <- unique(merged_data$Line)

# Merge chr_window_data for each line with the merged_data using a left join
for (line in unique_lines) {
  # Filter the merged_data for the current line
  line_data <- merged_data %>%
    filter(Line == line) %>%
    select(-Line)  # Remove the Line column from the current line data

  # Perform a left join to fill in the missing values and replace NA with NA_integer_
  merged_data_filled <- left_join(chr_window_data, line_data, by = "chr_Window") %>%
    replace_na(list(hotspot_no = NA_integer_))

  # Rename the hotspot_no column to include the line name as a prefix
  col_name <- paste0(line)
  col_names <- c("chr_Window", col_name)
  names(merged_data_filled)[2] <- col_name

  # Merge the current line data with the final data using the chr_Window column
  final_data <- left_join(final_data, merged_data_filled[, col_names], by = "chr_Window")
}


###################
# Plotting heatmap

# Transpose final_data and convert it to a matrix
final_data_matrix <- t(final_data[, -1])

# Create a vector for row names based on chr_Window column
row_names <- final_data[, 1]

# Remove chr_Window column from the matrix
colnames(final_data_matrix) <- row_names

# Define the colors for the heatmap
colors <- c("red1", "blue3", "purple1")
# Define the desired order for rows (lines)
desired_order <- rev(c("B97", "Oh43", "Il14H", "W22", "Mo18W", "Oh7B", "A344", "M162W", "CML322", 
                   "CML69", "Ki11", "CML228", "Mo17",  "CML277", "CML103"))
# Reorder the rows based on the desired order
final_data_matrix_reordered <- final_data_matrix[order(match(rownames(final_data_matrix), desired_order)), ]


# Create a vector with the data
chrwin <- c(1543, 2762, 3904, 5156, 6288, 7195, 8125, 9038, 9854, 10617)

# Convert the vector to a matrix with one row and ten columns
matrix_chrwin <- matrix(chrwin, nrow = 1)

# Add column names to the matrix
colnames(matrix_chrwin) <- c(1:10)

# Convert the matrix to a table
chrom_end_positions <- as.table(matrix_chrwin)

# Add row names to the table
rownames(chrom_end_positions) <- " "


# Create the heatmap using heatmap.2 with custom axis labels and no column names
heatmap(final_data_matrix_reordered, Colv = NA, Rowv = NA, scale = "none", col = colors,
        cexRow = 1.4, labCol = FALSE,  # Remove column names
        add.expr = {
          # Add vertical lines between groups of columns (chromosomes)
         # chrom_end_positions <- cumsum(table(sub("_.*", "", colnames(final_data_matrix_reordered))))
          for (pos in chrom_end_positions[-length(chrom_end_positions)]) {
            abline(v = pos + 0.5, col = "grey80", lty = "dotted", lwd = 0.7)
          }

          # Add chromosome numbers as axis labels for each group with the prefix "chr" and increased text size
          chrom_numbers <- as.character(seq_along(chrom_end_positions))
          axis(1, at = chrom_end_positions - (diff(c(0, chrom_end_positions))/2), 
               labels = paste0("chr", chrom_numbers), cex.axis = 1.5, lwd = 0.5,
               mgp = c(2.5, 1, 0))  

          # Add x-axis line covering from the beginning of chr 1 to the end of chr 10
          abline(h = c(0.5, sum(diff(chrom_end_positions)) + 0.5), col = "black", lwd = 1.1)

          # Add lines for each row (each line) with transparent fill
          for (row in 1:nrow(final_data_matrix_reordered)) {
            rect(0.5, row - 0.5, ncol(final_data_matrix_reordered) + 0.5, row + 0.5, 
                 col = rgb(0, 0, 0, alpha = 0), border= "grey80", lwd = 0.05)
          }
        })
# Add a custom legend to the left of the plot (adjust the 'x' and 'y' values as needed)
legend("top", legend = c("F", "M", "F & M"), fill = colors, border = NA, bty = "n", cex = 1.3, 
       x = 0, y = 0.68)



```


<a href="#top">Back to top</a>

<div style="margin-bottom:150px;">
</div>






```{r, echo=FALSE, message=FALSE, warning=FALSE, include= FALSE, dpi=300, fig.width=12, fig.height=8}
## 6) F vs M contrast spots for each chromosome
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher than the other sex \n among windows where recombination rate is at least more than half of the genome average.

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage")


contrastspot_files <- dir(pattern = '*_FM_contrastspot.csv$', recursive = TRUE)

# Create an empty list to store the data frames
contrastspotall_list <- list()

# Read the files and store them in the list along with the corresponding sample name
for (i in 1:length(contrastspot_files)) {
  df <- read.csv(contrastspot_files[i])
  # Extract the sample name from the file name
  sample_name <- sub("_FM_contrastspot\\.csv$", "", basename(contrastspot_files[i]))
  line <- substr(sample_name, 1, nchar(sample_name))
  # Add new columns with the sample information
  df$Line <- line
    # Add the hotspot_no column based on the Sex column
  df$contrastspot_no <- ifelse(df$FM_ratio > 0, 1, ifelse(df$FM_ratio < 0, 2, NA))

    # Store the data frame in the list
  contrastspotall_list[[i]] <- df
}

# Combine all the data frames into a single data frame
contrastspotall <- do.call(rbind, contrastspotall_list)

contrastspotall$chr_Window <- paste(contrastspotall$chr,contrastspotall$window, sep = "_")


# Remove duplicates if any (to ensure only one row per chr_Window and Line combination)
contrastspotall <- distinct(contrastspotall)


# Order the result by Line and chr_Window
contrastspotall <- contrastspotall %>%
  arrange(Line, chr_Window)

# Get the column names
column_names <- names(contrastspotall)

# Move column 10 to position 1
new_order <- c(column_names[10], column_names[-10])

# Reorder the data frame
contrastspotall <- contrastspotall[new_order]

##################


# Define the desired window ranges for each chromosome
chromosomes <- 1:10
window_ranges <- list(
  c(1:1543),
  c(1:1219),
  c(1:1142),
  c(1:1252),
  c(1:1132),
  c(1:907),
  c(1:930),
  c(1:913),
  c(1:816),
  c(1:763)
)

# Create a data frame with all chr_Window combinations based on the desired window ranges
chr_window_data <- data.frame(chr_Window = character(),
                              stringsAsFactors = FALSE)

for (chrom in chromosomes) {
  window_range <- window_ranges[[chrom]]
  chr_Window <- paste0(chrom, "_", window_range)
  chr_window_data <- rbind(chr_window_data, data.frame(chr_Window, stringsAsFactors = FALSE))
}

# Create a data frame to store the final merged data
contrast_final_data <- data.frame(chr_window_data, stringsAsFactors = FALSE)

# Get unique lines from the contrastspotall data
unique_lines <- unique(contrastspotall$Line)

# Merge chr_window_data for each line with the contrastspotall using a left join
for (line in unique_lines) {
  # Filter the contrastspotall for the current line
  line_data <- contrastspotall %>%
    filter(Line == line) %>%
    select(-Line)  # Remove the Line column from the current line data

  # Perform a left join to fill in the missing values and replace NA with NA_integer_
  contrastspotall_filled <- left_join(chr_window_data, line_data, by = "chr_Window") %>%
    replace_na(list(contrastspot_ratio = NA_integer_))

  # Rename the contrastspot_ratio column to include the line name as a prefix
  col_name <- paste0(line)
  col_names <- c("chr_Window", col_name)
  names(contrastspotall_filled)[2] <- col_name

  # Merge the current line data with the final data using the chr_Window column
  contrast_final_data <- left_join(contrast_final_data, contrastspotall_filled[, col_names], by = "chr_Window")
}


###################
# Plotting heatmap

# Transpose final_data and convert it to a matrix
contrast_final_data_matrix <- t(contrast_final_data[, -1])

# Create a vector for row names based on chr_Window column
row_names <- contrast_final_data[, 1]

# Remove chr_Window column from the matrix
colnames(contrast_final_data_matrix) <- row_names

# Define the colors for the heatmap
colors <- c("red1", "blue3")
# Define the desired order for rows (lines)
desired_order <- rev(c("B97", "Oh43", "Il14H", "W22", "Mo18W", "Oh7B", "A344", "M162W", "CML322", 
                   "CML69", "Ki11", "CML228", "Mo17",  "CML277", "CML103"))
# Reorder the rows based on the desired order
contrast_final_data_matrix_reordered <- contrast_final_data_matrix[order(match(rownames(contrast_final_data_matrix), desired_order)), ]

# Create a vector with the data
chrwin <- c(1543, 2762, 3904, 5156, 6288, 7195, 8125, 9038, 9854, 10617)

# Convert the vector to a matrix with one row and ten columns
matrix_chrwin <- matrix(chrwin, nrow = 1)

# Add column names to the matrix
colnames(matrix_chrwin) <- c(1:10)

# Convert the matrix to a table
chrom_end_positions <- as.table(matrix_chrwin)

# Add row names to the table
rownames(chrom_end_positions) <- " "


# Create the heatmap using heatmap.2 with custom axis labels and no column names
heatmap(contrast_final_data_matrix_reordered, Colv = NA, Rowv = NA, scale = "none", col = colors,
        cexRow = 1.4, labCol = FALSE,  # Remove column names
        add.expr = {
          # Add vertical lines between groups of columns (chromosomes)
          #chrom_end_positions <- cumsum(table(sub("_.*", "", colnames(contrast_final_data_matrix_reordered))))
          for (pos in chrom_end_positions[-length(chrom_end_positions)]) {
            abline(v = pos + 0.5, col = "grey80", lty = "dotted", lwd = 0.7)
          }

          # Add chromosome numbers as axis labels for each group with the prefix "chr" and increased text size
          chrom_numbers <- as.character(seq_along(chrom_end_positions))
          axis(1, at = chrom_end_positions - (diff(c(0, chrom_end_positions))/2), 
               labels = paste0("chr", chrom_numbers), cex.axis = 1.5, lwd = 0.5,
               mgp = c(2.5, 1, 0))  

          # Add x-axis line covering from the beginning of chr 1 to the end of chr 10
          abline(h = c(0.5, sum(diff(chrom_end_positions)) + 0.5), col = "black", lwd = 1.1)

          # Add lines for each row (each line) with transparent fill
          for (row in 1:nrow(contrast_final_data_matrix_reordered)) {
            rect(0.5, row - 0.5, ncol(contrast_final_data_matrix_reordered) + 0.5, row + 0.5, 
                 col = rgb(0, 0, 0, alpha = 0), border= "grey80", lwd = 0.05)
          }
        })
# Add a custom legend to the left of the plot (adjust the 'x' and 'y' values as needed)
legend("top", legend = c("F", "M"), fill = colors, border = NA, bty = "n", cex = 1.3, 
       x = 0, y = 0.68)

#####



```


<a href="#top">Back to top</a>

<div style="margin-bottom:150px;">
</div>



## 6) F vs M contrast spots among hotspots
##### *F vs M recombination contrast spots were determined also in 0.2Mb windows with at least 5-fold higher recombination than the other sex among hotspots.


<div style="margin-bottom:20px;">
</div>
```{r, echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=12, fig.height=10}

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage")


contrastspot_files <- dir(pattern = '*_FM_contrastspot.csv$', recursive = TRUE)

# Create an empty list to store the data frames
contrastspotall_list <- list()

# Read the files and store them in the list along with the corresponding sample name
for (i in 1:length(contrastspot_files)) {
  df <- read.csv(contrastspot_files[i])
  # Extract the sample name from the file name
  sample_name <- sub("_FM_contrastspot\\.csv$", "", basename(contrastspot_files[i]))
  line <- substr(sample_name, 1, nchar(sample_name))
  # Add new columns with the sample information
  df$Line <- line
  df$general_hotspot <- ifelse(df$general_hotspot == "FALSE", "F", df$general_hotspot)
    # Add the hotspot_no column based on the Sex column
  df$contrastspot_no <- ifelse(df$general_hotspot == "F", 1, ifelse(df$general_hotspot == "M", 2, NA))

    # Store the data frame in the list
  contrastspotall_list[[i]] <- df
}

# Combine all the data frames into a single data frame
contrastspotall_hot <- do.call(rbind, contrastspotall_list)



contrastspotall_hot <- subset(contrastspotall_hot, contrastspot_no == 1 | contrastspot_no == 2)

contrastspotall_hot$chr_Window <- paste(contrastspotall_hot$chr,contrastspotall_hot$window, sep = "_")


# Remove duplicates if any (to ensure only one row per chr_Window and Line combination)
contrastspotall_hot <- distinct(contrastspotall_hot)


# Order the result by Line and chr_Window
contrastspotall_hot <- contrastspotall_hot %>%
  arrange(Line, chr_Window)

# Get the column names
column_names_hot <- names(contrastspotall_hot)

# Move column 10 to position 1
new_order_hot <- c(column_names_hot[10], column_names_hot[-10])

# Reorder the data frame
contrastspotall_hot <- contrastspotall_hot[new_order_hot]

##################


# Define the desired window ranges for each chromosome
chromosomes <- 1:10

window_ranges <- list(
  c(1:1543),
  c(1:1219),
  c(1:1142),
  c(1:1252),
  c(1:1132),
  c(1:907),
  c(1:930),
  c(1:913),
  c(1:816),
  c(1:763)
)

# Create a data frame with all chr_Window combinations based on the desired window ranges
chr_window_data <- data.frame(chr_Window = character(),
                              stringsAsFactors = FALSE)

for (chrom in chromosomes) {
  window_range <- window_ranges[[chrom]]
  chr_Window <- paste0(chrom, "_", window_range)
  chr_window_data <- rbind(chr_window_data, data.frame(chr_Window, stringsAsFactors = FALSE))
}

# Create a data frame to store the final merged data
contrast_hot_final_data <- data.frame(chr_window_data, stringsAsFactors = FALSE)

# Get unique lines from the contrastspotall data
unique_lines_hot <- unique(contrastspotall_hot$Line)

# Merge chr_window_data for each line with the contrastspotall using a left join
for (line in unique_lines_hot) {
  # Filter the contrastspotall for the current line
  line_data <- contrastspotall_hot %>%
    filter(Line == line) %>%
    select(-Line)  # Remove the Line column from the current line data

  # Perform a left join to fill in the missing values and replace NA with NA_integer_
  contrastspotall_hot_filled <- left_join(chr_window_data, line_data, by = "chr_Window") %>%
    replace_na(list(contrastspot_ratio = NA_integer_))

  # Rename the contrastspot_ratio column to include the line name as a prefix
  col_name <- paste0(line)
  col_names <- c("chr_Window", col_name)
  names(contrastspotall_hot_filled)[2] <- col_name

  # Merge the current line data with the final data using the chr_Window column
  contrast_hot_final_data <- left_join(contrast_hot_final_data, contrastspotall_hot_filled[, col_names], by = "chr_Window")
}

# Add missing lines
contrast_hot_final_data$A344 <- NA
contrast_hot_final_data$CML103 <- NA


###################
# Plotting heatmap

# Transpose final_data and convert it to a matrix
contrast_hot_final_data_matrix <- t(contrast_hot_final_data[, -1])

# Create a vector for row names based on chr_Window column
row_names <- contrast_hot_final_data[, 1]

# Remove chr_Window column from the matrix
colnames(contrast_hot_final_data_matrix) <- row_names

# Define the colors for the heatmap
colors <- c("red1", "blue3")
# Define the desired order for rows (lines)
desired_order <- rev(c("B97", "Oh43", "Il14H", "W22", "Mo18W", "Oh7B", "A344", "M162W", "CML322", 
                   "CML69", "Ki11", "CML228", "Mo17",  "CML277", "CML103"))
# Reorder the rows based on the desired order
contrast_hot_final_data_matrix_reordered <- contrast_hot_final_data_matrix[order(match(rownames(contrast_hot_final_data_matrix), desired_order)), ]


# Create a vector with the data
chrwin <- c(1543, 2762, 3904, 5156, 6288, 7195, 8125, 9038, 9854, 10617)

# Convert the vector to a matrix with one row and ten columns
matrix_chrwin <- matrix(chrwin, nrow = 1)

# Add column names to the matrix
colnames(matrix_chrwin) <- c(1:10)

# Convert the matrix to a table
chrom_end_positions <- as.table(matrix_chrwin)

# Add row names to the table
rownames(chrom_end_positions) <- " "




# Create the heatmap using heatmap.2 with custom axis labels and no column names
heatmap(contrast_hot_final_data_matrix_reordered, Colv = NA, Rowv = NA, scale = "none", col = colors,
        cexRow = 1.4, labCol = FALSE,  # Remove column names
        add.expr = {
          # Add vertical lines between groups of columns (chromosomes)
         # chrom_end_positions <- cumsum(table(sub("_.*", "", colnames(contrast_hot_final_data_matrix_reordered))))
          for (pos in chrom_end_positions[-length(chrom_end_positions)]) {
            abline(v = pos + 0.5, col = "grey80", lty = "dotted", lwd = 0.7)
          }

          # Add chromosome numbers as axis labels for each group with the prefix "chr" and increased text size
          chrom_numbers <- as.character(seq_along(chrom_end_positions))
          axis(1, at = chrom_end_positions - (diff(c(0, chrom_end_positions))/2), 
               labels = paste0("chr", chrom_numbers), cex.axis = 1.5, lwd = 0.5,
               mgp = c(2.5, 1, 0))  

          # Add x-axis line covering from the beginning of chr 1 to the end of chr 10
          abline(h = c(0.5, sum(diff(chrom_end_positions)) + 0.5), col = "black", lwd = 1.1)

          # Add lines for each row (each line) with transparent fill
          for (row in 1:nrow(contrast_hot_final_data_matrix_reordered)) {
            rect(0.5, row - 0.5, ncol(contrast_hot_final_data_matrix_reordered) + 0.5, row + 0.5, 
                 col = rgb(0, 0, 0, alpha = 0), border= "grey80", lwd = 0.05)
          }
        })




#####


# Bar plot

A344_NA <- c(NA,NA,NA,NA,NA,NA,NA,NA,"NA","A344",NA)
CML103_NA <- c(NA,NA,NA,NA,NA,NA,NA,NA,"NA","CML103",NA)

contrastspotall_hotall <- rbind(contrastspotall_hot, A344_NA,CML103_NA)


# Reorder levels for Line variable
contrastspotall_hotall$Line <- factor(contrastspotall_hotall$Line, 
                                      levels = c("B97", "Oh43", "Il14H", "W22", "Mo18W", "Oh7B", 
                                                 "A344", "M162W", "CML322", "CML69", "Ki11","CML228",
                                                 "Mo17", "CML277", "CML103"))


contrastspotall_hotall$general_hotspot <- factor(contrastspotall_hotall$general_hotspot, 
                                      levels = c( "M", "F", "NA"))



bar_contrast <- ggplot() + 
  geom_bar(data = contrastspotall_hotall, aes(x = Line, fill = general_hotspot), 
           position = "stack") +
  theme_bw() +
  scale_fill_manual(values=c("blue3", "red1", "white")) +
  labs(y="Number of contrast spot in hotspots", title="", x="BC1s crossed with B73") +
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), 
          axis.title=element_text(size=14),
          legend.position = "none", 
          plot.title = element_blank(),
          panel.grid.major = element_blank(), panel.grid.minor = element_blank()) 


png(filename = "plot/Contrastspot_bar.png", width = 10, height = 5, units = "in",  res = 400 )
bar_contrast
dev.off()





```


<a href="#top">Back to top</a>

<div style="margin-bottom:150px;">
</div>



```{r, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=10, fig.height=4}

## 4) Different variation of CO numbers between sexes

#F vs M

COall.aov <- aov( CO ~ line*sex, data = COall_sum)
#summary(COall.aov)
#TukeyHSD(COall.aov)
COall_res <- HSD.test(COall.aov, c("line", "sex"), group=TRUE)
#COall_res$groups
oder.group_COall <- COall_res$groups[order(rownames(COall_res$groups)),] 
#oder.group_COall


COall_sum$line <- factor(COall_sum$line, levels = c("B97", "Oh43", "Il14H", "W22", "Oh7B","A344", "CML277","M162W", "Mo18W",
                                                    "CML322", "CML69", "Ki11", "Mo17",  "CML103"))


COall_sum_sex <- ggplot(COall_sum, aes(x= line, y = CO, fill = sex)) +
  geom_violin(trim = T, bw = 1, alpha =0.7) +
   stat_summary(fun = mean,
               geom = "pointrange",
               fun.max = function(x) mean(x) + sd(x),
               fun.min = function(x) mean(x) - sd(x), 
              color= "white", size= 0.5, linewidth = 0.5) +
  theme_classic() +
   facet_wrap( ~ sex, ncol = 9, strip.position = "bottom") +
    labs(y="Number of crossover", title="", x="BC1s crossed with B73") +
    theme(axis.text.y=element_text(size=10), axis.text.x=element_blank(), axis.ticks.x=element_blank(), 
          strip.background = element_blank(), strip.placement = "outside", 
          axis.title=element_text(size=14), strip.text = element_text(size=12, margin = margin(0,0,20,0)), 
          legend.position = "bottom",  legend.justification = "right",
          legend.title=element_blank(), legend.margin = margin(-27, 0, 0, 0),
          plot.title = element_blank() ) +
    scale_fill_manual(values=c("red2", "blue4"),
                      breaks=c("F", "M"))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


COall_sum_sex

  
```





## 7) Genetic distance acorss lines
<div style="margin-bottom:20px;">
</div>
```{r, echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=8, fig.height=7}

library(readr)
library(dplyr)
library(stringr)
library(ggplot2)
library(circlize)

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout")

peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")



# Get list of file names in directory
marey_names <- list.files(pattern = "*mareyout.chr_snp.txt$", recursive = TRUE)

# Read in files and add unique identifier column
data_list <- lapply(marey_names, function(file) {
  data <- read_delim(file, delim = " ", col_names = TRUE)
  data$file_name <- file
  data
})

# Bind rows and remove duplicate rows
merged_mareyout <- bind_rows(data_list) %>%
               dplyr::select(-file_name) %>%
               distinct()

# Add sexes
merged_mareyout <- merged_mareyout %>%
   dplyr::mutate(sex = str_sub(set, -1))

merged_mareyout$Mb <- merged_mareyout$phys/1000000

# View merged data
#head(merged_mareyout)

colnames(merged_mareyout)[2] <- "chr" 

merged_mareyout$chr <- as.numeric(sub("Chromosome ", "", merged_mareyout$chr))

merged_mareyout$line <- gsub("F$|M$", "", merged_mareyout$set)

merged_mareyout <- merged_mareyout %>%
  select(set, line, sex, chr, mkr, phys, gen, vld, spline, Mb)

# Save the merged and updated file 
##write.csv(merged_mareyout, file = "merged_mareyout.csv", row.names = FALSE, quote = FALSE)


## Plotting cM

ALL_CM <- ggplot() +
    geom_rect( data=peri, aes(xmin=Start_Mb, xmax=End_Mb), ymin=-Inf, ymax=Inf, 
    alpha = 0.3, fill = 'grey') +
    geom_point(data =merged_mareyout, aes(x= Mb, y= gen, col = sex), alpha =0.02, lwd =0.2, size = 0.8) +
    facet_wrap( ~ chr, ncol = 2, scales = "free") +
    theme_bw() +
    labs(x= "Physical positions (Mb)" ,
         y="Genetic distance (cM) of all lines", title="") +
    theme(axis.text.y=element_text(size=8),
          axis.text.x=element_text(size=12),
          axis.title=element_text(size=15),
          strip.text = element_text(size=12),
          strip.background = element_blank(),
          strip.text.y.left = element_text(angle = 0),
          legend.text = element_text(size=12),
          legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),) +
    scale_color_manual(values=c('red3','blue4'), name = "") +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
    guides(colour = guide_legend(override.aes = list(size=2, alpha=0.8)))


ALL_CM


```
<a href="#top">Back to top</a>

<div style="margin-bottom:150px;">
</div>



## 8) Variation of lines in recombination rates  
<div style="margin-bottom:20px;">
</div>

```{r, echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=8, fig.height=7}
# sd per line per sex per chr 
library(readr)
library(dplyr)
library(stringr)
library(ggplot2)
library(circlize)

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout/")

peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")



# Get list of file names in directory
marey_names <- list.files(pattern = "*mareyout.chr_snp.txt$", recursive = TRUE)

# Read in files and add unique identifier column
data_list <- lapply(marey_names, function(file) {
  data <- read_delim(file, delim = " ", col_names = TRUE)
  data$file_name <- file
  data
})

# Bind rows and remove duplicate rows
merged_mareyout <- bind_rows(data_list) %>%
               select(-file_name) %>%
               distinct()

# Add sexes
merged_mareyout <- merged_mareyout %>%
   dplyr::mutate(sex = str_sub(set, -1))

merged_mareyout$Mb <- merged_mareyout$phys/1000000

# View merged data
#head(merged_mareyout)

colnames(merged_mareyout)[2] <- "chr" 

merged_mareyout$chr <- as.numeric(sub("Chromosome ", "", merged_mareyout$chr))

merged_mareyout$line <- gsub("F$|M$", "", merged_mareyout$set)

merged_mareyout <- merged_mareyout %>%
  select(set, line, sex, chr, mkr, phys, gen, vld, spline, Mb)


############################

# First, filter the data to keep only the variables needed for the plot
plot_data <- merged_mareyout %>%
  select(line, sex, chr, phys, spline)

# Then, create bins of 1 Mb for the physical position column
plot_data$Mb_bin <- cut(plot_data$phys, breaks = seq(0, max(plot_data$phys), by = 1e6))

# Calculate the midpoint of each Mb_bin interval
plot_data$Mb <- (as.numeric(gsub("\\((.*),.*\\]", "\\1", plot_data$Mb_bin)) + 500000)/1e+06

# Then, calculate the average spline for each line within each 1 Mb bin, grouped by sex
plot_data1 <- plot_data %>%
  dplyr::group_by(line, sex, chr, Mb_bin, Mb) %>%
  dplyr::summarise(avg_spline = mean(spline, na.rm = TRUE))



# Then, calculate the SD of the average spline for each line within each 1 Mb bin, grouped by sex
plot_data2 <- plot_data1 %>%
  dplyr::group_by(sex, chr, Mb_bin, Mb) %>%
  dplyr::summarise(SD_avg_spline = sd(avg_spline, na.rm = TRUE))

plot_data2$sex <- factor(plot_data2$sex, levels = c("F", "M"))


# Finally, create the plot using ggplot2
SD_ALL <- ggplot() +
  # Add the area plot layer
  geom_area(data = plot_data2, aes(x = Mb, y = SD_avg_spline, color = sex, fill = sex, group = sex),
            position = position_dodge(), alpha = 0.3, lwd = 0.5) +
#  geom_smooth(data = plot_data2, aes(x = Mb_bin, y = sd_avg_spline, color = sex, group = sex),
#              method = "spline", se = FALSE, size = 0.4, span = 0.05, alpha = 0.9) + # add smoothed line
  # Add the rectangle layer
  geom_rect(data = peri, aes(xmin = Start_Mb, xmax = End_Mb, ymin = -Inf, ymax = Inf), fill = "gray80", alpha = 0.3) +
  # Add facet wrap and axis labels
  facet_wrap(~chr, ncol = 2, scales = "free") +
  labs(x = "Physical positions (Mb)", y = "Standard deviation of recombination rates acorss lines", title = "") +
  theme_bw() +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 10),
        axis.title = element_text(size = 14),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 10),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4')) 




SD_ALL

 
```
<a href="#top">Back to top</a>



<div style="margin-bottom:150px;">
</div>



## 9) CV of recombination rates
##### *Pericentromeric regions were excluded because of too low mean recombination rates and high variation
<div style="margin-bottom:15px;">
</div>

```{r, echo=FALSE, message=FALSE, warning=FALSE, dpi=300, fig.width=8, fig.height=7}
library(readr)
library(dplyr)
library(stringr)
library(ggplot2)
library(circlize)

setwd("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/linkage/mareyout/")

peri <- read.csv("C:/Users/Gwonjin Lee/OneDrive - University of Florida/Project/Maize/GBS_analysis/CO/ZmB73V5_heterochromatin regions.csv")



# Get list of file names in directory
marey_names <- list.files(pattern = "*mareyout.chr_snp.txt$", recursive = TRUE)

# Read in files and add unique identifier column
data_list <- lapply(marey_names, function(file) {
  data <- read_delim(file, delim = " ", col_names = TRUE)
  data$file_name <- file
  data
})

# Bind rows and remove duplicate rows
merged_mareyout <- bind_rows(data_list) %>%
               select(-file_name) %>%
               distinct()

# Add sexes
merged_mareyout <- merged_mareyout %>%
   dplyr::mutate(sex = str_sub(set, -1))

merged_mareyout$Mb <- merged_mareyout$phys/1000000

# View merged data
#head(merged_mareyout)

colnames(merged_mareyout)[2] <- "chr" 

merged_mareyout$chr <- as.numeric(sub("Chromosome ", "", merged_mareyout$chr))

merged_mareyout$line <- gsub("F$|M$", "", merged_mareyout$set)

merged_mareyout <- merged_mareyout %>%
  select(set, line, sex, chr, mkr, phys, gen, vld, spline, Mb)

# First, filter the data to keep only the variables needed for the plot
plot_data_cv <- merged_mareyout %>%
  select(line, sex, chr, phys, spline)

# Then, create bins of 1 Mb for the physical position column
plot_data_cv$Mb_bin <- cut(plot_data_cv$phys, breaks = seq(0, max(plot_data_cv$phys), by = 1e6))

# Calculate the midpoint of each Mb_bin interval
plot_data_cv$Mb <- (as.numeric(gsub("\\((.*),.*\\]", "\\1", plot_data_cv$Mb_bin)) + 500000) / 1e+06

# Then, calculate the average spline for each line within each 1 Mb bin, grouped by sex
plot_data_cv1 <- plot_data_cv %>%
  dplyr::group_by(line, sex, chr, Mb_bin, Mb) %>%
  dplyr::summarise(avg_spline = mean(spline, na.rm = TRUE))

# Convert avg_spline to 0 if it is less than 0
plot_data_cv1$avg_spline[plot_data_cv1$avg_spline < 0] <- 0

# Then, calculate the standard deviation of the average spline for each line within each 1 Mb bin, grouped by sex
# plot_data_cv <- plot_data_cv %>%
#   group_by(sex, chr, Mb_bin, Mb) %>%
#   summarise(sd_avg_spline = sd(avg_spline, na.rm = TRUE))

# Then, calculate the coefficient of variation (CV) of the average spline for each line within each 1 Mb bin, grouped by sex
plot_data_cv2 <- plot_data_cv1 %>%
  dplyr::group_by(sex, chr, Mb_bin, Mb) %>%
  dplyr::summarise(cv_avg_spline = (sd(avg_spline, na.rm = TRUE) / mean(avg_spline, na.rm = TRUE)))


plot_data_cv2_filtered <- plot_data_cv2 %>%
  dplyr::left_join(peri, by = "chr") %>%
  dplyr::filter(!(Mb >= Start_Mb & Mb <= End_Mb)) %>%
  dplyr::select(-Start, -End, -Start_Mb, -End_Mb)


cv_ALL <- ggplot() +
  # Add the area plot layer
    geom_area(data = plot_data_cv2, aes(x = Mb, y = cv_avg_spline, color = sex, fill = sex, group = sex),
            position = position_dodge(), alpha = 0.3, lwd = 0.5) +
 # geom_line(data = plot_data_cv2_filtered, aes(x = Mb, y = cv_avg_spline, color = sex, group = sex),
#            alpha = 0.6, size = 1.2) +
  # Add facet wrap and axis labels
  facet_wrap(~chr, ncol = 2, scales = "free") +
  # Add the rectangle layer (placed after geom_line, so it will be drawn on top)
  geom_rect(data = peri, aes(xmin = Start_Mb, xmax = End_Mb, ymin = -Inf, ymax = Inf),
            fill = "gray95", alpha = 1) +
  labs(x = "Physical positions (Mb)", y = "Coefficient of variation of recombination rates acorss lines", title = "") +
  theme_bw() + ylim(0,2.5) +
  theme(axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 10),
        axis.title = element_text(size = 14),
        strip.text = element_text(size = 12),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0),
        legend.text = element_text(size = 10),
        legend.title = element_blank(),
        legend.position = "none",  # adjust these values to position the legend where you want
        legend.justification = c(1, 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), plot.title = element_text(hjust = 0.5)) +
  scale_color_manual(values=c('red2','blue4')) +
  scale_fill_manual(values=c('red2','blue4'))





cv_ALL




 
```
<a href="#top">Back to top</a>



<div style="margin-bottom:150px;">
</div>

